// @tsparticles/interaction-external-connect/tsparticles.interaction.external.connect.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine'), require('@tsparticles/canvas-utils')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine', '@tsparticles/canvas-utils'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalConnect = global.__tsParticlesInternals.interactions.externalConnect || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine, global.__tsParticlesInternals.canvas.utils));
})(this, (function (exports, pluginInteractivity, engine, canvasUtils) { 'use strict';

    class ConnectLinks {
        opacity;
        constructor() {
            this.opacity = 0.5;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.opacity !== undefined) {
                this.opacity = data.opacity;
            }
        }
    }

    class Connect {
        distance;
        links;
        radius;
        constructor() {
            this.distance = 80;
            this.links = new ConnectLinks();
            this.radius = 60;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.distance !== undefined) {
                this.distance = data.distance;
            }
            this.links.load(data.links);
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
        }
    }

    const gradientMin = 0, gradientMax = 1, defaultLinksWidth = 0;
    function gradient(container, context, p1, p2, opacity) {
        const gradStop = Math.floor(p2.getRadius() / p1.getRadius()), color1 = p1.getFillColor(), color2 = p2.getFillColor();
        if (!color1 || !color2) {
            return;
        }
        const sourcePos = p1.getPosition(), destPos = p2.getPosition(), midRgb = engine.colorMix(color1, color2, p1.getRadius(), p2.getRadius()), grad = context.createLinearGradient(sourcePos.x, sourcePos.y, destPos.x, destPos.y);
        grad.addColorStop(gradientMin, engine.getStyleFromHsl(color1, container.hdr, opacity));
        grad.addColorStop(engine.clamp(gradStop, gradientMin, gradientMax), engine.getStyleFromRgb(midRgb, container.hdr, opacity));
        grad.addColorStop(gradientMax, engine.getStyleFromHsl(color2, container.hdr, opacity));
        return grad;
    }
    function drawConnectLine(context, width, lineStyle, begin, end) {
        canvasUtils.drawLine(context, begin, end);
        context.lineWidth = width;
        context.strokeStyle = lineStyle;
        context.stroke();
    }
    function lineStyle(container, ctx, p1, p2) {
        const options = container.actualOptions, connectOptions = options.interactivity?.modes.connect;
        if (!connectOptions) {
            return;
        }
        return gradient(container, ctx, p1, p2, connectOptions.links.opacity);
    }
    function drawConnection(container, p1, p2) {
        container.canvas.render.draw(ctx => {
            const ls = lineStyle(container, ctx, p1, p2);
            if (!ls) {
                return;
            }
            const pos1 = p1.getPosition(), pos2 = p2.getPosition();
            drawConnectLine(ctx, p1.retina.linksWidth ?? defaultLinksWidth, ls, pos1, pos2);
        });
    }

    const connectMode = "connect", minDistance = 0;
    class Connector extends pluginInteractivity.ExternalInteractorBase {
        #maxDistance;
        constructor(container) {
            super(container);
            this.#maxDistance = 0;
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear() {
        }
        init() {
            const container = this.container, connect = container.actualOptions.interactivity?.modes.connect;
            if (!connect) {
                return;
            }
            this.#maxDistance = connect.distance;
            container.retina.connectModeDistance = connect.distance * container.retina.pixelRatio;
            container.retina.connectModeRadius = connect.radius * container.retina.pixelRatio;
        }
        interact(interactivityData) {
            const container = this.container, options = container.actualOptions;
            if (options.interactivity?.events.onHover.enable && interactivityData.status === "pointermove") {
                const mousePos = interactivityData.mouse.position, { connectModeDistance, connectModeRadius } = container.retina;
                if (!connectModeDistance ||
                    connectModeDistance < minDistance ||
                    !connectModeRadius ||
                    connectModeRadius < minDistance ||
                    !mousePos) {
                    return;
                }
                const distance = Math.abs(connectModeRadius), query = container.particles.grid.queryCircle(mousePos, distance, p => this.isEnabled(interactivityData, p));
                query.forEach((p1, i) => {
                    const pos1 = p1.getPosition(), indexOffset = 1;
                    for (const p2 of query.slice(i + indexOffset)) {
                        const pos2 = p2.getPosition(), distMax = Math.abs(connectModeDistance), xDiff = Math.abs(pos1.x - pos2.x), yDiff = Math.abs(pos1.y - pos2.y);
                        if (xDiff < distMax && yDiff < distMax) {
                            drawConnection(container, p1, p2);
                        }
                    }
                });
            }
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
            if (!(events?.onHover.enable && mouse.position)) {
                return false;
            }
            return engine.isInArray(connectMode, events.onHover.mode);
        }
        loadModeOptions(options, ...sources) {
            options.connect ??= new Connect();
            for (const source of sources) {
                options.connect.load(source?.connect);
            }
        }
        reset() {
        }
    }

    async function loadExternalConnectInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalConnect", container => {
                return Promise.resolve(new Connector(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalConnectInteraction = loadExternalConnectInteraction;

    exports.Connect = Connect;
    exports.ConnectLinks = ConnectLinks;
    exports.loadExternalConnectInteraction = loadExternalConnectInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalConnectInteraction: (globalThis.__tsParticlesInternals.interactions.externalConnect || {}).loadExternalConnectInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
