// @tsparticles/engine/cjs/Utils/EventDispatcher.js
import { deleteCount, minIndex } from "../Core/Utils/Constants.js";
export class EventDispatcher {
    #listeners;
    constructor() {
        this.#listeners = new Map();
    }
    addEventListener(type, listener) {
        this.removeEventListener(type, listener);
        let arr = this.#listeners.get(type);
        if (!arr) {
            arr = [];
            this.#listeners.set(type, arr);
        }
        arr.push(listener);
    }
    dispatchEvent(type, args) {
        const listeners = this.#listeners.get(type);
        listeners?.forEach(handler => {
            handler(args);
        });
    }
    hasEventListener(type) {
        return !!this.#listeners.get(type);
    }
    removeAllEventListeners(type) {
        if (!type) {
            this.#listeners = new Map();
        }
        else {
            this.#listeners.delete(type);
        }
    }
    removeEventListener(type, listener) {
        const arr = this.#listeners.get(type);
        if (!arr) {
            return;
        }
        const length = arr.length, idx = arr.indexOf(listener);
        if (idx < minIndex) {
            return;
        }
        if (length === deleteCount) {
            this.#listeners.delete(type);
        }
        else {
            arr.splice(idx, deleteCount);
        }
    }
}
