// @tsparticles/plugin-blend/tsparticles.plugin.blend.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.blend = global.__tsParticlesInternals.plugins.blend || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class Blend {
        enable;
        mode;
        constructor() {
            this.mode = "destination-out";
            this.enable = false;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.mode !== undefined) {
                this.mode = data.mode;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
        }
    }

    class BlendPlugin {
        id = "blend";
        async getPlugin(container) {
            const { BlendPluginInstance } = await Promise.resolve().then(function () { return BlendPluginInstance$1; });
            return new BlendPluginInstance(container);
        }
        loadOptions(_container, options, source) {
            if (!this.needsPlugin(options) && !this.needsPlugin(source)) {
                return;
            }
            let blendOptions = options.blend;
            if (!blendOptions?.load) {
                options.blend = blendOptions = new Blend();
            }
            blendOptions.load(source?.blend);
        }
        loadParticlesOptions(_container, options, source) {
            options.blend ??= new Blend();
            options.blend.load(source?.blend);
        }
        needsPlugin(options) {
            return !!options?.blend?.enable || !!options?.particles?.blend?.enable;
        }
    }

    async function loadBlendPlugin(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register(e => {
            e.pluginManager.addPlugin(new BlendPlugin());
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadBlendPlugin = loadBlendPlugin;

    class BlendPluginInstance {
        #container;
        #defaultCompositeValue;
        constructor(container) {
            this.#container = container;
        }
        drawParticleCleanup(context, particle) {
            if (!particle.options.blend?.enable) {
                return;
            }
            context.globalCompositeOperation = particle.originalBlendMode ?? engine.defaultCompositeValue;
            particle.originalBlendMode = undefined;
        }
        drawParticleSetup(context, particle) {
            if (!particle.options.blend?.enable) {
                return;
            }
            particle.originalBlendMode = context.globalCompositeOperation;
            context.globalCompositeOperation = particle.options.blend.mode;
        }
        drawSettingsCleanup(context) {
            if (!this.#defaultCompositeValue) {
                return;
            }
            context.globalCompositeOperation = this.#defaultCompositeValue;
        }
        drawSettingsSetup(context) {
            const previousComposite = context.globalCompositeOperation, blend = this.#container.actualOptions.blend;
            this.#defaultCompositeValue = previousComposite;
            context.globalCompositeOperation = blend?.enable ? blend.mode : previousComposite;
        }
    }

    var BlendPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        BlendPluginInstance: BlendPluginInstance
    });

    exports.loadBlendPlugin = loadBlendPlugin;

}));
Object.assign(globalThis.window || globalThis, { loadBlendPlugin: (globalThis.__tsParticlesInternals.plugins.blend || {}).loadBlendPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
