// @tsparticles/engine/esm/Utils/LogUtils.js
const errorPrefix = "tsParticles - Error";
const wrap = (fn) => (...args) => {
    fn(...args);
}, _logger = {
    debug: wrap(console.debug),
    error: (message, ...optionalParams) => {
        console.error(`${errorPrefix} - ${message}`, ...optionalParams);
    },
    info: wrap(console.info),
    log: wrap(console.log),
    trace: wrap(console.trace),
    verbose: wrap(console.log),
    warning: wrap(console.warn),
};
export function setLogger(logger) {
    if (logger.debug) {
        _logger.debug = wrap(logger.debug);
    }
    if (logger.error) {
        _logger.error = wrap(logger.error);
    }
    if (logger.info) {
        _logger.info = wrap(logger.info);
    }
    if (logger.log) {
        _logger.log = wrap(logger.log);
    }
    if (logger.trace) {
        _logger.trace = wrap(logger.trace);
    }
    if (logger.verbose) {
        _logger.verbose = wrap(logger.verbose);
    }
    if (logger.warning) {
        _logger.warning = wrap(logger.warning);
    }
}
export function getLogger() {
    return _logger;
}
