// @tsparticles/interaction-particles-collisions/tsparticles.interaction.particles.collisions.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Particles Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.particlesCollisions = global.__tsParticlesInternals.interactions.particlesCollisions || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    var CollisionMode;
    (function (CollisionMode) {
        CollisionMode["absorb"] = "absorb";
        CollisionMode["bounce"] = "bounce";
        CollisionMode["destroy"] = "destroy";
    })(CollisionMode || (CollisionMode = {}));

    class CollisionsAbsorb {
        speed;
        constructor() {
            this.speed = 2;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.speed !== undefined) {
                this.speed = data.speed;
            }
        }
    }

    class CollisionsOverlap {
        enable;
        retries;
        constructor() {
            this.enable = true;
            this.retries = 0;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.retries !== undefined) {
                this.retries = data.retries;
            }
        }
    }

    class Collisions {
        absorb;
        bounce;
        enable;
        maxSpeed;
        mode;
        overlap;
        constructor() {
            this.absorb = new CollisionsAbsorb();
            this.bounce = new engine.ParticlesBounce();
            this.enable = false;
            this.maxSpeed = 50;
            this.mode = CollisionMode.bounce;
            this.overlap = new CollisionsOverlap();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            this.absorb.load(data.absorb);
            this.bounce.load(data.bounce);
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.maxSpeed !== undefined) {
                this.maxSpeed = engine.setRangeValue(data.maxSpeed);
            }
            if (data.mode !== undefined) {
                this.mode = data.mode;
            }
            this.overlap.load(data.overlap);
        }
    }

    const minAbsorbAmount = 0;
    function updateAbsorb(p1, r1, p2, r2, delta, pixelRatio) {
        if (!p1.options.collisions || !p2.options.collisions) {
            return;
        }
        const absorbSpeed = p1.options.collisions.absorb.speed, shrinkAmount = engine.clamp(absorbSpeed * delta.factor, minAbsorbAmount, r2);
        p1.size.value = Math.sqrt(r1 * r1 + shrinkAmount * shrinkAmount);
        p2.size.value -= shrinkAmount;
        if (p2.size.value <= pixelRatio) {
            p2.size.value = 0;
            p2.destroy();
        }
    }
    function absorb(p1, p2, delta, pixelRatio) {
        const r1 = p1.getRadius(), r2 = p2.getRadius();
        if (!r1 && r2) {
            p1.destroy();
        }
        else if (r1 && !r2) {
            p2.destroy();
        }
        else if (r1 && r2) {
            if (r1 >= r2) {
                updateAbsorb(p1, r1, p2, r2, delta, pixelRatio);
            }
            else {
                updateAbsorb(p2, r2, p1, r1, delta, pixelRatio);
            }
        }
    }

    const energyCorrectionMinRatio = 1e-6, energyDriftThreshold = 1e-4, correctionFactorNeutral = 1, fixBounceSpeed = (p) => {
        if (!p.options.collisions) {
            return;
        }
        p.collisionMaxSpeed ??= engine.getRangeValue(p.options.collisions.maxSpeed);
        if (p.velocity.length > p.collisionMaxSpeed) {
            p.velocity.length = p.collisionMaxSpeed;
        }
    };
    function bounce(p1, p2) {
        const m1 = p1.getMass(), m2 = p2.getMass(), speed1Before = p1.velocity.length, speed2Before = p2.velocity.length, keBefore = m1 * speed1Before * speed1Before + m2 * speed2Before * speed2Before;
        engine.circleBounce(engine.circleBounceDataFromParticle(p1), engine.circleBounceDataFromParticle(p2));
        const speed1After = p1.velocity.length, speed2After = p2.velocity.length, keAfter = m1 * speed1After * speed1After + m2 * speed2After * speed2After;
        if (keAfter > keBefore * energyCorrectionMinRatio) {
            const correctionFactor = Math.sqrt(keBefore / keAfter);
            if (Math.abs(correctionFactor - correctionFactorNeutral) > energyDriftThreshold) {
                p1.velocity.length = speed1After * correctionFactor;
                p2.velocity.length = speed2After * correctionFactor;
            }
        }
        fixBounceSpeed(p1);
        fixBounceSpeed(p2);
    }

    function destroy(p1, p2) {
        if (!p1.unbreakable && !p2.unbreakable) {
            bounce(p1, p2);
        }
        const p1Radius = p1.getRadius(), p2Radius = p2.getRadius();
        if (!p1Radius && p2Radius) {
            p1.destroy();
        }
        else if (p1Radius && !p2Radius) {
            p2.destroy();
        }
        else if (p1Radius && p2Radius) {
            const deleteP = p1.getRadius() >= p2.getRadius() ? p2 : p1;
            deleteP.destroy();
        }
    }

    function resolveCollision(p1, p2, delta, pixelRatio) {
        if (!p1.options.collisions || !p2.options.collisions) {
            return;
        }
        switch (p1.options.collisions.mode) {
            case CollisionMode.absorb: {
                absorb(p1, p2, delta, pixelRatio);
                break;
            }
            case CollisionMode.bounce: {
                bounce(p1, p2);
                break;
            }
            case CollisionMode.destroy: {
                destroy(p1, p2);
                break;
            }
        }
    }

    class Collider extends pluginInteractivity.ParticlesInteractorBase {
        maxDistance;
        constructor(container) {
            super(container);
            this.maxDistance = 0;
        }
        clear() {
        }
        init() {
        }
        interact(p1, _interactivityData, delta) {
            if (p1.destroyed || p1.spawning) {
                return;
            }
            const container = this.container, pos1 = p1.getPosition(), radius1 = p1.getRadius(), query = container.particles.grid.queryCircle(pos1, radius1 * engine.double);
            for (const p2 of query) {
                if (p1 === p2 ||
                    p1.id >= p2.id ||
                    !p1.options.collisions?.enable ||
                    !p2.options.collisions?.enable ||
                    p1.options.collisions.mode !== p2.options.collisions.mode ||
                    p2.destroyed ||
                    p2.spawning) {
                    continue;
                }
                const pos2 = p2.getPosition(), radius2 = p2.getRadius();
                if (Math.abs(Math.round(pos1.z) - Math.round(pos2.z)) > radius1 + radius2) {
                    continue;
                }
                const dist = engine.getDistance(pos1, pos2), distP = radius1 + radius2;
                if (dist > distP) {
                    continue;
                }
                resolveCollision(p1, p2, delta, container.retina.pixelRatio);
            }
        }
        isEnabled(particle) {
            return !!particle.options.collisions?.enable;
        }
        loadParticlesOptions(options, ...sources) {
            options.collisions ??= new Collisions();
            for (const source of sources) {
                options.collisions.load(source?.collisions);
            }
        }
        reset() {
        }
    }

    class OverlapPlugin {
        id = "overlap";
        async getPlugin(container) {
            const { OverlapPluginInstance } = await Promise.resolve().then(function () { return OverlapPluginInstance$1; });
            return new OverlapPluginInstance(container);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }

    async function loadParticlesCollisionsInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addPlugin(new OverlapPlugin());
            e.pluginManager.addInteractor?.("particlesCollisions", container => {
                return Promise.resolve(new Collider(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadParticlesCollisionsInteraction = loadParticlesCollisionsInteraction;

    const minRetries = 0;
    class OverlapPluginInstance {
        #container;
        constructor(container) {
            this.#container = container;
        }
        checkParticlePosition(particle, position, tryCount) {
            return !this.#hasOverlaps(particle, position, tryCount);
        }
        #hasOverlaps = (particle, pos, tryCount) => {
            const collisionsOptions = particle.options.collisions;
            if (!collisionsOptions?.enable) {
                return false;
            }
            const overlapOptions = collisionsOptions.overlap;
            if (overlapOptions.enable) {
                return false;
            }
            const retries = overlapOptions.retries;
            if (retries >= minRetries && tryCount > retries) {
                throw new Error(`Particle is overlapping and can't be placed`);
            }
            return !!this.#container.particles.find(p => engine.getDistance(pos, p.position) < particle.getRadius() + p.getRadius());
        };
    }

    var OverlapPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        OverlapPluginInstance: OverlapPluginInstance
    });

    exports.loadParticlesCollisionsInteraction = loadParticlesCollisionsInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadParticlesCollisionsInteraction: (globalThis.__tsParticlesInternals.interactions.particlesCollisions || {}).loadParticlesCollisionsInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
