// @tsparticles/interaction-external-parallax/tsparticles.interaction.external.parallax.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalParallax = global.__tsParticlesInternals.interactions.externalParallax || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class Parallax {
        force;
        smooth;
        constructor() {
            this.force = 2;
            this.smooth = 10;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.force !== undefined) {
                this.force = data.force;
            }
            if (data.smooth !== undefined) {
                this.smooth = data.smooth;
            }
        }
    }

    const parallaxMode = "parallax";
    class Parallaxer extends pluginInteractivity.ExternalInteractorBase {
        maxDistance = 0;
        constructor(container) {
            super(container);
        }
        clear() {
        }
        init() {
        }
        interact(interactivityData) {
            for (const particle of this.container.particles.filter(p => this.isEnabled(interactivityData, p))) {
                this.#parallaxInteract(interactivityData, particle);
            }
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
            return !!events?.onHover.enable && !!mouse.position && engine.isInArray(parallaxMode, events.onHover.mode);
        }
        loadModeOptions(options, ...sources) {
            options.parallax ??= new Parallax();
            for (const source of sources) {
                options.parallax.load(source?.parallax);
            }
        }
        reset() {
        }
        #parallaxInteract(interactivityData, particle) {
            if (!this.isEnabled(interactivityData, particle)) {
                return;
            }
            const container = this.container, options = container.actualOptions, parallaxOptions = options.interactivity?.modes.parallax;
            if (!parallaxOptions) {
                return;
            }
            const parallaxForce = parallaxOptions.force, mousePos = interactivityData.mouse.position;
            if (!mousePos) {
                return;
            }
            const canvasSize = container.canvas.size, canvasCenter = {
                x: canvasSize.width * engine.half,
                y: canvasSize.height * engine.half,
            }, parallaxSmooth = parallaxOptions.smooth, factor = particle.getRadius() / parallaxForce, centerDistance = {
                x: (mousePos.x - canvasCenter.x) * factor,
                y: (mousePos.y - canvasCenter.y) * factor,
            }, { offset } = particle;
            offset.x += (centerDistance.x - offset.x) / parallaxSmooth;
            offset.y += (centerDistance.y - offset.y) / parallaxSmooth;
        }
    }

    async function loadExternalParallaxInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalParallax", container => {
                return Promise.resolve(new Parallaxer(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalParallaxInteraction = loadExternalParallaxInteraction;

    exports.Parallax = Parallax;
    exports.loadExternalParallaxInteraction = loadExternalParallaxInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalParallaxInteraction: (globalThis.__tsParticlesInternals.interactions.externalParallax || {}).loadExternalParallaxInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
