// @tsparticles/canvas-utils/tsparticles.canvas.utils.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Utility v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.canvas = global.__tsParticlesInternals.canvas || {}, global.__tsParticlesInternals.canvas.utils = global.__tsParticlesInternals.canvas.utils || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    const defaultWidth = 0, indexesOffset = {
        r: 0,
        g: 1,
        b: 2,
        a: 3,
    }, alphaMax = 255, alphaFactor = engine.identity / alphaMax;
    function drawLine(context, begin, end) {
        context.beginPath();
        context.moveTo(begin.x, begin.y);
        context.lineTo(end.x, end.y);
        context.closePath();
    }
    function getCanvasImageData(ctx, size, offset, clear = true) {
        const imageData = ctx.getImageData(engine.originPoint.x, engine.originPoint.y, size.width, size.height).data;
        if (clear) {
            ctx.clearRect(engine.originPoint.x, engine.originPoint.y, size.width, size.height);
        }
        const pixels = [];
        for (let i = 0; i < imageData.length; i += offset) {
            const idx = i / offset, pos = {
                x: idx % size.width,
                y: Math.floor(idx / size.width),
            };
            pixels[pos.y] ??= [];
            const row = pixels[pos.y];
            if (!row) {
                continue;
            }
            row[pos.x] = {
                r: imageData[i + indexesOffset.r] ?? engine.defaultRgbMin,
                g: imageData[i + indexesOffset.g] ?? engine.defaultRgbMin,
                b: imageData[i + indexesOffset.b] ?? engine.defaultRgbMin,
                a: (imageData[i + indexesOffset.a] ?? engine.defaultAlpha) * alphaFactor,
            };
        }
        return {
            pixels,
            width: Math.min(...pixels.map(row => row.length)),
            height: pixels.length,
        };
    }
    function getImageData(src, offset, canvasSettings) {
        const image = new Image();
        image.crossOrigin = "Anonymous";
        const p = new Promise((resolve, reject) => {
            image.onerror = reject;
            image.onload = () => {
                const canvas = engine.safeDocument().createElement("canvas");
                canvas.width = image.width;
                canvas.height = image.height;
                const context = canvas.getContext("2d", canvasSettings);
                if (!context) {
                    reject(new Error("Could not get canvas context"));
                    return;
                }
                context.drawImage(image, engine.originPoint.x, engine.originPoint.y, image.width, image.height, engine.originPoint.x, engine.originPoint.y, canvas.width, canvas.height);
                resolve(getCanvasImageData(context, canvas, offset));
            };
        });
        image.src = src;
        return p;
    }
    function getTextData(textOptions, offset, fill, canvasSettings) {
        const canvas = engine.safeDocument().createElement("canvas"), context = canvas.getContext("2d", canvasSettings), { font, text, lines: linesOptions, color } = textOptions;
        if (!text || !context) {
            return;
        }
        const lines = text.split(linesOptions.separator), fontSize = engine.isNumber(font.size) ? `${font.size.toString()}px` : font.size, linesData = [];
        let maxWidth = 0, totalHeight = 0;
        for (const line of lines) {
            context.font = `${font.style || ""} ${font.variant || ""} ${font.weight || ""} ${fontSize} ${font.family}`;
            const measure = context.measureText(line), lineData = {
                measure,
                text: line,
                height: measure.actualBoundingBoxAscent + measure.actualBoundingBoxDescent,
                width: measure.width,
            };
            maxWidth = Math.max(maxWidth || defaultWidth, lineData.width);
            totalHeight += lineData.height + linesOptions.spacing;
            linesData.push(lineData);
        }
        canvas.width = maxWidth;
        canvas.height = totalHeight;
        let currentHeight = 0;
        for (const line of linesData) {
            context.font = `${font.style || ""} ${font.variant || ""} ${font.weight || ""} ${fontSize} ${font.family}`;
            if (fill) {
                context.fillStyle = color;
                context.fillText(line.text, engine.originPoint.x, currentHeight + line.measure.actualBoundingBoxAscent);
            }
            else {
                context.strokeStyle = color;
                context.strokeText(line.text, engine.originPoint.x, currentHeight + line.measure.actualBoundingBoxAscent);
            }
            currentHeight += line.height + linesOptions.spacing;
        }
        return getCanvasImageData(context, canvas, offset);
    }
    async function loadFont(font, weight) {
        try {
            await engine.safeDocument().fonts.load(`${weight ?? "400"} 36px '${font ?? "Verdana"}'`);
        }
        catch {
        }
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};

    exports.drawLine = drawLine;
    exports.getCanvasImageData = getCanvasImageData;
    exports.getImageData = getImageData;
    exports.getTextData = getTextData;
    exports.loadFont = loadFont;

}));
Object.assign(globalThis.window || globalThis, { loadFont: (globalThis.__tsParticlesInternals.canvas.utils || {}).loadFont });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
