// @tsparticles/interaction-external-destroy/tsparticles.interaction.external.destroy.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalDestroy = global.__tsParticlesInternals.interactions.externalDestroy || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    const half = 0.5, toleranceFactor = 10, minRadius = 0;
    function processDestroy(container, area) {
        const query = container.particles.grid.query(area);
        for (const particle of query) {
            particle.destroy();
        }
    }
    function singleSelectorDestroy(container, selector, div) {
        const query = engine.safeDocument().querySelectorAll(selector);
        if (!query.length) {
            return;
        }
        query.forEach(item => {
            const elem = item, pxRatio = container.retina.pixelRatio, pos = {
                x: (elem.offsetLeft + elem.offsetWidth * half) * pxRatio,
                y: (elem.offsetTop + elem.offsetHeight * half) * pxRatio,
            }, radius = elem.offsetWidth * half * pxRatio, tolerance = toleranceFactor * pxRatio, area = div.type === pluginInteractivity.DivType.circle
                ? new engine.Circle(pos.x, pos.y, radius + tolerance)
                : new engine.Rectangle(elem.offsetLeft * pxRatio - tolerance, elem.offsetTop * pxRatio - tolerance, elem.offsetWidth * pxRatio + tolerance * engine.double, elem.offsetHeight * pxRatio + tolerance * engine.double);
            processDestroy(container, area);
        });
    }
    function divDestroy(container, divs, destroyMode) {
        pluginInteractivity.divModeExecute(destroyMode, divs, (selector, div) => {
            singleSelectorDestroy(container, selector, div);
        });
    }
    function mouseDestroy(container, interactivityData) {
        const pxRatio = container.retina.pixelRatio, tolerance = toleranceFactor * pxRatio, mousePos = interactivityData.mouse.position, radius = container.retina.destroyModeDistance;
        if (!radius || radius < minRadius || !mousePos) {
            return;
        }
        processDestroy(container, new engine.Circle(mousePos.x, mousePos.y, radius + tolerance));
    }

    class Destroy {
        distance;
        constructor() {
            this.distance = 200;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.distance !== undefined) {
                this.distance = data.distance;
            }
        }
    }

    const destroyMode = "destroy";
    class Destroyer extends pluginInteractivity.ExternalInteractorBase {
        #maxDistance;
        constructor(container) {
            super(container);
            this.#maxDistance = 0;
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear() {
        }
        init() {
            const container = this.container, destroy = container.actualOptions.interactivity?.modes.destroy;
            if (!destroy) {
                return;
            }
            this.#maxDistance = destroy.distance;
            container.retina.destroyModeDistance = destroy.distance * container.retina.pixelRatio;
        }
        interact(interactivityData) {
            const container = this.container, options = container.actualOptions, events = options.interactivity?.events, mouseMoveStatus = interactivityData.status === pluginInteractivity.mouseMoveEvent;
            if (!events) {
                return;
            }
            const hoverEnabled = events.onHover.enable, hoverMode = events.onHover.mode, divs = events.onDiv;
            if (mouseMoveStatus && hoverEnabled && engine.isInArray(destroyMode, hoverMode)) {
                mouseDestroy(this.container, interactivityData);
            }
            else {
                divDestroy(this.container, divs, destroyMode);
            }
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
            if (!events) {
                return false;
            }
            const divs = events.onDiv;
            return ((!!mouse.position && events.onHover.enable && engine.isInArray(destroyMode, events.onHover.mode)) ||
                pluginInteractivity.isDivModeEnabled(destroyMode, divs));
        }
        loadModeOptions(options, ...sources) {
            options.destroy ??= new Destroy();
            for (const source of sources) {
                options.destroy.load(source?.destroy);
            }
        }
        reset() {
        }
    }

    async function loadExternalDestroyInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalDestroy", async (container) => {
                return Promise.resolve(new Destroyer(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalDestroyInteraction = loadExternalDestroyInteraction;

    exports.Destroy = Destroy;
    exports.loadExternalDestroyInteraction = loadExternalDestroyInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalDestroyInteraction: (globalThis.__tsParticlesInternals.interactions.externalDestroy || {}).loadExternalDestroyInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
