// @tsparticles/interaction-particles-collisions/browser/Collider.js
import { double, getDistance } from "@tsparticles/engine";
import { ParticlesInteractorBase } from "@tsparticles/plugin-interactivity";
import { Collisions } from "./Options/Classes/Collisions.js";
import { resolveCollision } from "./ResolveCollision.js";
export class Collider extends ParticlesInteractorBase {
    maxDistance;
    constructor(container) {
        super(container);
        this.maxDistance = 0;
    }
    clear() {
    }
    init() {
    }
    interact(p1, _interactivityData, delta) {
        if (p1.destroyed || p1.spawning) {
            return;
        }
        const container = this.container, pos1 = p1.getPosition(), radius1 = p1.getRadius(), query = container.particles.grid.queryCircle(pos1, radius1 * double);
        for (const p2 of query) {
            if (p1 === p2 ||
                p1.id >= p2.id ||
                !p1.options.collisions?.enable ||
                !p2.options.collisions?.enable ||
                p1.options.collisions.mode !== p2.options.collisions.mode ||
                p2.destroyed ||
                p2.spawning) {
                continue;
            }
            const pos2 = p2.getPosition(), radius2 = p2.getRadius();
            if (Math.abs(Math.round(pos1.z) - Math.round(pos2.z)) > radius1 + radius2) {
                continue;
            }
            const dist = getDistance(pos1, pos2), distP = radius1 + radius2;
            if (dist > distP) {
                continue;
            }
            resolveCollision(p1, p2, delta, container.retina.pixelRatio);
        }
    }
    isEnabled(particle) {
        return !!particle.options.collisions?.enable;
    }
    loadParticlesOptions(options, ...sources) {
        options.collisions ??= new Collisions();
        for (const source of sources) {
            options.collisions.load(source?.collisions);
        }
    }
    reset() {
    }
}
