// @tsparticles/interaction-external-remove/cjs/Remover.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { getRangeValue, isInArray } from "@tsparticles/engine";
import { Remove } from "./Options/Classes/Remove.js";
const removeMode = "remove";
export class Remover extends ExternalInteractorBase {
    handleClickMode;
    maxDistance = 0;
    constructor(container) {
        super(container);
        this.handleClickMode = (mode) => {
            const container = this.container, options = container.actualOptions;
            if (!options.interactivity?.modes.remove || mode !== removeMode) {
                return;
            }
            const removeNb = getRangeValue(options.interactivity.modes.remove.quantity);
            for (let i = 0; i < removeNb; i++) {
                container.particles.get(i)?.destroy();
            }
        };
    }
    clear() {
    }
    init() {
    }
    interact() {
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
        return !!events && mouse.clicking && mouse.inside && !!mouse.position && isInArray(removeMode, events.onClick.mode);
    }
    loadModeOptions(options, ...sources) {
        options.remove ??= new Remove();
        for (const source of sources) {
            options.remove.load(source?.remove);
        }
    }
    reset() {
    }
}
