// @tsparticles/shape-image/browser/ImageDrawer.js
import { defaultAlpha, defaultRatio, double, } from "@tsparticles/engine";
import { replaceImageColor, shapeTypes } from "./Utils.js";
import { drawGif } from "./GifUtils/Utils.js";
const sides = 12;
export class ImageDrawer {
    #container;
    #engine;
    constructor(engine, container) {
        this.#engine = engine;
        this.#container = container;
    }
    draw(data) {
        const { context, radius, particle, opacity } = data, image = particle.image, element = image?.element;
        if (!image) {
            return;
        }
        context.globalAlpha = opacity;
        if (image.gif && image.gifData) {
            drawGif(data, this.#container.canvas.render.settings);
        }
        else if (element) {
            const ratio = image.ratio, pos = {
                x: -radius,
                y: -radius,
            }, diameter = radius * double;
            context.drawImage(element, pos.x, pos.y, diameter, diameter / ratio);
        }
        context.globalAlpha = defaultAlpha;
    }
    getSidesCount() {
        return sides;
    }
    async init(container) {
        const options = container.actualOptions;
        if (!options.preload || !this.#engine.loadImage) {
            return;
        }
        const promises = [];
        for (const imageData of options.preload) {
            promises.push(this.#engine.loadImage(container, imageData));
        }
        await Promise.all(promises);
    }
    loadShape(particle) {
        const container = this.#container;
        if (!particle.shape || !shapeTypes.includes(particle.shape)) {
            return;
        }
        const imageData = particle.shapeData;
        if (!imageData) {
            return;
        }
        const images = this.#engine.getImages?.(container), image = images?.find((t) => t.name === imageData.name || t.source === imageData.src);
        if (image) {
            return;
        }
        void this.#loadImageShape(container, imageData).then(() => {
            this.loadShape(particle);
        });
    }
    particleInit(container, particle) {
        if (particle.shape !== "image" && particle.shape !== "images") {
            return;
        }
        const images = this.#engine.getImages?.(container), imageData = particle.shapeData;
        if (!imageData) {
            return;
        }
        const color = particle.getFillColor(), image = images?.find((t) => t.name === imageData.name || t.source === imageData.src);
        if (!image) {
            return;
        }
        const replaceColor = imageData.replaceColor;
        if (image.loading) {
            setTimeout(() => {
                this.particleInit(container, particle);
            });
            return;
        }
        void (async () => {
            let imageRes;
            if (image.svgData && color) {
                imageRes = await replaceImageColor(image, imageData, color, particle, container.hdr);
            }
            else {
                imageRes = {
                    color,
                    data: image,
                    element: image.element,
                    gif: image.gif,
                    gifData: image.gifData,
                    gifLoopCount: image.gifLoopCount,
                    loaded: true,
                    ratio: imageData.width && imageData.height ? imageData.width / imageData.height : (image.ratio ?? defaultRatio),
                    replaceColor: replaceColor,
                    source: imageData.src,
                };
            }
            if (!imageRes.ratio) {
                imageRes.ratio = 1;
            }
            const close = imageData.close ?? particle.shapeClose, imageShape = {
                image: imageRes,
                close,
            };
            particle.image = imageShape.image;
            particle.shapeClose = imageShape.close;
        })();
    }
    #loadImageShape = async (container, imageShape) => {
        if (!this.#engine.loadImage) {
            throw new Error(`Image shape not initialized`);
        }
        await this.#engine.loadImage(container, {
            gif: imageShape.gif,
            name: imageShape.name,
            replaceColor: imageShape.replaceColor,
            src: imageShape.src,
        });
    };
}
