// @tsparticles/interaction-external-parallax/esm/Parallaxer.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { half, isInArray } from "@tsparticles/engine";
import { Parallax } from "./Options/Classes/Parallax.js";
const parallaxMode = "parallax";
export class Parallaxer extends ExternalInteractorBase {
    maxDistance = 0;
    constructor(container) {
        super(container);
    }
    clear() {
    }
    init() {
    }
    interact(interactivityData) {
        for (const particle of this.container.particles.filter(p => this.isEnabled(interactivityData, p))) {
            this.#parallaxInteract(interactivityData, particle);
        }
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
        return !!events?.onHover.enable && !!mouse.position && isInArray(parallaxMode, events.onHover.mode);
    }
    loadModeOptions(options, ...sources) {
        options.parallax ??= new Parallax();
        for (const source of sources) {
            options.parallax.load(source?.parallax);
        }
    }
    reset() {
    }
    #parallaxInteract(interactivityData, particle) {
        if (!this.isEnabled(interactivityData, particle)) {
            return;
        }
        const container = this.container, options = container.actualOptions, parallaxOptions = options.interactivity?.modes.parallax;
        if (!parallaxOptions) {
            return;
        }
        const parallaxForce = parallaxOptions.force, mousePos = interactivityData.mouse.position;
        if (!mousePos) {
            return;
        }
        const canvasSize = container.canvas.size, canvasCenter = {
            x: canvasSize.width * half,
            y: canvasSize.height * half,
        }, parallaxSmooth = parallaxOptions.smooth, factor = particle.getRadius() / parallaxForce, centerDistance = {
            x: (mousePos.x - canvasCenter.x) * factor,
            y: (mousePos.y - canvasCenter.y) * factor,
        }, { offset } = particle;
        offset.x += (centerDistance.x - offset.x) / parallaxSmooth;
        offset.y += (centerDistance.y - offset.y) / parallaxSmooth;
    }
}
