// @tsparticles/plugin-interactivity/cjs/InteractivityEventListeners.js
import { double, executeOnSingleOrMultiple, lengthOffset, manageListener, safeDocument, visibilityChangeEvent, } from "@tsparticles/engine";
import { mouseDownEvent, mouseLeaveEvent, mouseMoveEvent, mouseUpEvent, touchCancelEvent, touchEndEvent, touchMoveEvent, touchStartEvent, } from "./InteractivityConstants.js";
import { InteractivityDetect } from "./Enums/InteractivityDetect.js";
const touchDelay = 500;
export class InteractivityEventListeners {
    #canPush = true;
    #clickPositionPlugins;
    #container;
    #handlers;
    #interactionManager;
    #touches;
    constructor(container, interactionManager) {
        this.#container = container;
        this.#clickPositionPlugins = [];
        this.#interactionManager = interactionManager;
        this.#touches = new Map();
        this.#handlers = {
            mouseDown: () => {
                this.#mouseDown();
            },
            mouseLeave: () => {
                this.#mouseTouchFinish();
            },
            mouseMove: (e) => {
                this.#mouseTouchMove(e);
            },
            mouseUp: (e) => {
                this.#mouseTouchClick(e);
            },
            touchStart: (e) => {
                this.#touchStart(e);
            },
            touchMove: (e) => {
                this.#mouseTouchMove(e);
            },
            touchEnd: (e) => {
                this.#touchEnd(e);
            },
            touchCancel: (e) => {
                this.#touchEnd(e);
            },
            touchEndClick: (e) => {
                this.#touchEndClick(e);
            },
            visibilityChange: () => {
                this.#handleVisibilityChange();
            },
        };
    }
    addListeners() {
        this.#manageListeners(true);
    }
    init() {
        this.#clickPositionPlugins.length = 0;
        for (const plugin of this.#container.plugins.filter(p => !!p.clickPositionValid)) {
            this.#clickPositionPlugins.push(plugin);
        }
    }
    removeListeners() {
        this.#manageListeners(false);
    }
    #doMouseTouchClick = e => {
        const container = this.#container, interactionManager = this.#interactionManager, options = container.actualOptions;
        if (this.#canPush) {
            const mouseInteractivity = interactionManager.interactivityData.mouse, mousePos = mouseInteractivity.position;
            if (!mousePos) {
                return;
            }
            mouseInteractivity.clickPosition = { ...mousePos };
            mouseInteractivity.clickTime = performance.now();
            const onClick = options.interactivity?.events.onClick;
            if (!onClick?.mode) {
                return;
            }
            executeOnSingleOrMultiple(onClick.mode, mode => {
                interactionManager.handleClickMode(mode);
            });
        }
        if (e.type === "touchend") {
            setTimeout(() => {
                this.#mouseTouchFinish();
            }, touchDelay);
        }
    };
    #handleVisibilityChange = () => {
        this.#mouseTouchFinish();
    };
    #manageInteractivityListeners = add => {
        const handlers = this.#handlers, container = this.#container, interactionManager = this.#interactionManager, options = container.actualOptions, interactivityEl = interactionManager.interactivityData.element;
        if (!interactivityEl) {
            return;
        }
        const html = interactivityEl, canvas = container.canvas;
        canvas.setPointerEvents(html === canvas.domElement ? "initial" : "none");
        if (add && !(options.interactivity?.events.onHover.enable || options.interactivity?.events.onClick.enable)) {
            return;
        }
        manageListener(interactivityEl, mouseMoveEvent, handlers.mouseMove, add);
        manageListener(interactivityEl, touchStartEvent, handlers.touchStart, add);
        manageListener(interactivityEl, touchMoveEvent, handlers.touchMove, add);
        if (add) {
            if (options.interactivity?.events.onClick.enable) {
                manageListener(interactivityEl, touchEndEvent, handlers.touchEndClick, add);
                manageListener(interactivityEl, mouseUpEvent, handlers.mouseUp, add);
                manageListener(interactivityEl, mouseDownEvent, handlers.mouseDown, add);
            }
            else {
                manageListener(interactivityEl, touchEndEvent, handlers.touchEnd, add);
            }
        }
        else {
            manageListener(interactivityEl, touchEndEvent, handlers.touchEndClick, add);
            manageListener(interactivityEl, mouseUpEvent, handlers.mouseUp, add);
            manageListener(interactivityEl, mouseDownEvent, handlers.mouseDown, add);
            manageListener(interactivityEl, touchEndEvent, handlers.touchEnd, add);
        }
        manageListener(interactivityEl, mouseLeaveEvent, handlers.mouseLeave, add);
        manageListener(interactivityEl, touchCancelEvent, handlers.touchCancel, add);
    };
    #manageListeners = add => {
        const handlers = this.#handlers, container = this.#container, interactionManager = this.#interactionManager, options = container.actualOptions, detectType = options.interactivity?.detectsOn, canvasEl = container.canvas.domElement;
        if (detectType === InteractivityDetect.window) {
            interactionManager.interactivityData.element = safeDocument();
        }
        else if (detectType === InteractivityDetect.parent && canvasEl) {
            interactionManager.interactivityData.element = canvasEl.parentElement ?? canvasEl.parentNode;
        }
        else {
            interactionManager.interactivityData.element = canvasEl;
        }
        this.#manageInteractivityListeners(add);
        manageListener(document, visibilityChangeEvent, handlers.visibilityChange, add, false);
    };
    #mouseDown = () => {
        const { interactivityData } = this.#interactionManager, { mouse } = interactivityData;
        mouse.clicking = true;
        mouse.downPosition = mouse.position;
    };
    #mouseTouchClick = e => {
        const container = this.#container, interactionManager = this.#interactionManager, options = container.actualOptions, { mouse } = interactionManager.interactivityData;
        mouse.inside = true;
        let handled = false;
        const mousePosition = mouse.position;
        if (!mousePosition || !options.interactivity?.events.onClick.enable) {
            return;
        }
        for (const plugin of this.#clickPositionPlugins) {
            handled = plugin.clickPositionValid?.(mousePosition) ?? false;
            if (handled) {
                break;
            }
        }
        if (!handled) {
            this.#doMouseTouchClick(e);
        }
        mouse.clicking = false;
    };
    #mouseTouchFinish = () => {
        const { interactivityData } = this.#interactionManager, { mouse } = interactivityData;
        delete mouse.position;
        delete mouse.clickPosition;
        delete mouse.downPosition;
        interactivityData.status = mouseLeaveEvent;
        mouse.inside = false;
        mouse.clicking = false;
    };
    #mouseTouchMove = e => {
        const container = this.#container, interactionManager = this.#interactionManager, options = container.actualOptions, interactivity = interactionManager.interactivityData, canvasEl = container.canvas.domElement;
        if (!interactivity.element) {
            return;
        }
        interactivity.mouse.inside = true;
        let pos;
        if (e.type.startsWith("pointer")) {
            this.#canPush = true;
            const mouseEvent = e;
            if (interactivity.element === safeDocument()) {
                if (canvasEl) {
                    const clientRect = canvasEl.getBoundingClientRect();
                    pos = {
                        x: mouseEvent.clientX - clientRect.left,
                        y: mouseEvent.clientY - clientRect.top,
                    };
                }
            }
            else if (options.interactivity?.detectsOn === InteractivityDetect.parent) {
                const source = mouseEvent.target, target = mouseEvent.currentTarget;
                if (canvasEl) {
                    const sourceRect = source.getBoundingClientRect(), targetRect = target.getBoundingClientRect(), canvasRect = canvasEl.getBoundingClientRect();
                    pos = {
                        x: mouseEvent.offsetX + double * sourceRect.left - (targetRect.left + canvasRect.left),
                        y: mouseEvent.offsetY + double * sourceRect.top - (targetRect.top + canvasRect.top),
                    };
                }
                else {
                    pos = {
                        x: mouseEvent.offsetX,
                        y: mouseEvent.offsetY,
                    };
                }
            }
            else if (mouseEvent.target === canvasEl) {
                pos = {
                    x: mouseEvent.offsetX,
                    y: mouseEvent.offsetY,
                };
            }
        }
        else {
            this.#canPush = e.type !== "touchmove";
            if (canvasEl) {
                const touchEvent = e, lastTouch = touchEvent.touches[touchEvent.touches.length - lengthOffset], canvasRect = canvasEl.getBoundingClientRect();
                if (!lastTouch) {
                    return;
                }
                pos = {
                    x: lastTouch.clientX - canvasRect.left,
                    y: lastTouch.clientY - canvasRect.top,
                };
            }
        }
        const pxRatio = container.retina.pixelRatio;
        if (pos) {
            pos.x *= pxRatio;
            pos.y *= pxRatio;
        }
        interactivity.mouse.position = pos;
        interactivity.status = mouseMoveEvent;
    };
    #touchEnd = e => {
        const evt = e, touches = Array.from(evt.changedTouches);
        for (const touch of touches) {
            this.#touches.delete(touch.identifier);
        }
        this.#mouseTouchFinish();
    };
    #touchEndClick = e => {
        const evt = e, touches = Array.from(evt.changedTouches);
        for (const touch of touches) {
            this.#touches.delete(touch.identifier);
        }
        this.#mouseTouchClick(e);
    };
    #touchStart = e => {
        const evt = e, touches = Array.from(evt.changedTouches);
        for (const touch of touches) {
            this.#touches.set(touch.identifier, performance.now());
        }
        this.#mouseTouchMove(e);
    };
}
