// @tsparticles/plugin-interactivity/cjs/InteractivityPluginInstance.js
import { InteractionManager } from "./InteractionManager.js";
import { Interactivity } from "./Options/Classes/Interactivity.js";
export class InteractivityPluginInstance {
    interactionManager;
    #container;
    #pluginManager;
    constructor(pluginManager, container) {
        this.#container = container;
        this.#pluginManager = pluginManager;
        this.interactionManager = new InteractionManager(pluginManager, container);
        this.#container.addClickHandler = (callback) => {
            this.interactionManager.addClickHandler(callback);
        };
    }
    addClickHandler(callback) {
        this.interactionManager.addClickHandler(callback);
    }
    clearClickHandlers() {
        this.interactionManager.clearClickHandlers();
    }
    destroy() {
        this.clearClickHandlers();
        this.#pluginManager.interactors?.delete(this.#container);
    }
    particleCreated(particle) {
        const interactivityParticle = particle, interactivity = new Interactivity(this.#pluginManager, this.#container);
        interactivity.load(this.#container.actualOptions.interactivity);
        interactivity.load(interactivityParticle.options.interactivity);
        interactivityParticle.interactivity = interactivity;
    }
    particleReset(particle) {
        this.interactionManager.reset(particle);
    }
    postParticleUpdate(particle, delta) {
        this.interactionManager.particlesInteract(particle, delta);
    }
    postUpdate(delta) {
        this.interactionManager.externalInteract(delta);
        this.interactionManager.updateMaxDistance();
    }
    async preInit() {
        await this.interactionManager.initInteractors();
        this.interactionManager.init();
    }
    async redrawInit() {
        await this.interactionManager.initInteractors();
        this.interactionManager.init();
    }
    start() {
        this.interactionManager.addListeners();
        this.interactionManager.startObserving();
        return Promise.resolve();
    }
    stop() {
        this.interactionManager.removeListeners();
        this.interactionManager.stopObserving();
    }
}
