// @tsparticles/interaction-particles-links/tsparticles.interaction.particles.links.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Particles Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.particlesLinks = global.__tsParticlesInternals.interactions.particlesLinks || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class CircleWarp extends engine.Circle {
        #canvasSize;
        constructor(x, y, radius, canvasSize) {
            super(x, y, radius);
            this.#canvasSize = canvasSize;
        }
        contains(point) {
            if (super.contains(point))
                return true;
            const { width, height } = this.#canvasSize, { x, y } = point;
            return (super.contains({ x: x - width, y }) ||
                super.contains({ x: x + width, y }) ||
                super.contains({ x, y: y - height }) ||
                super.contains({ x, y: y + height }) ||
                super.contains({ x: x - width, y: y - height }) ||
                super.contains({ x: x + width, y: y + height }) ||
                super.contains({ x: x - width, y: y + height }) ||
                super.contains({ x: x + width, y: y - height }));
        }
        intersects(range) {
            if (super.intersects(range))
                return true;
            const { width, height } = this.#canvasSize, pos = range.position, shifts = [
                { x: -width, y: 0 },
                { x: width, y: 0 },
                { x: 0, y: -height },
                { x: 0, y: height },
                { x: -width, y: -height },
                { x: width, y: height },
                { x: -width, y: height },
                { x: width, y: -height },
            ];
            for (const shift of shifts) {
                const shiftedPos = { x: pos.x + shift.x, y: pos.y + shift.y };
                let shiftedRange;
                if (range instanceof engine.Circle) {
                    shiftedRange = new engine.Circle(shiftedPos.x, shiftedPos.y, range.radius);
                }
                else {
                    const rect = range;
                    shiftedRange = new engine.Rectangle(shiftedPos.x, shiftedPos.y, rect.size.width, rect.size.height);
                }
                if (super.intersects(shiftedRange))
                    return true;
            }
            return false;
        }
    }

    class LinksShadow {
        blur;
        color;
        enable;
        constructor() {
            this.blur = 5;
            this.color = new engine.OptionsColor();
            this.color.value = "#000";
            this.enable = false;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.blur !== undefined) {
                this.blur = data.blur;
            }
            this.color = engine.OptionsColor.create(this.color, data.color);
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
        }
    }

    class LinksTriangle {
        color;
        enable;
        frequency;
        opacity;
        constructor() {
            this.enable = false;
            this.frequency = 1;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.color !== undefined) {
                this.color = engine.OptionsColor.create(this.color, data.color);
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.frequency !== undefined) {
                this.frequency = data.frequency;
            }
            if (data.opacity !== undefined) {
                this.opacity = data.opacity;
            }
        }
    }

    class Links {
        blink;
        color;
        consent;
        distance;
        enable;
        frequency;
        id;
        opacity;
        shadow;
        triangles;
        warp;
        width;
        constructor() {
            this.blink = false;
            this.color = new engine.OptionsColor();
            this.color.value = "#fff";
            this.consent = false;
            this.distance = 100;
            this.enable = false;
            this.frequency = 1;
            this.opacity = 1;
            this.shadow = new LinksShadow();
            this.triangles = new LinksTriangle();
            this.width = 1;
            this.warp = false;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.id !== undefined) {
                this.id = data.id;
            }
            if (data.blink !== undefined) {
                this.blink = data.blink;
            }
            this.color = engine.OptionsColor.create(this.color, data.color);
            if (data.consent !== undefined) {
                this.consent = data.consent;
            }
            if (data.distance !== undefined) {
                this.distance = data.distance;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.frequency !== undefined) {
                this.frequency = data.frequency;
            }
            if (data.opacity !== undefined) {
                this.opacity = data.opacity;
            }
            this.shadow.load(data.shadow);
            this.triangles.load(data.triangles);
            if (data.width !== undefined) {
                this.width = data.width;
            }
            if (data.warp !== undefined) {
                this.warp = data.warp;
            }
        }
    }

    const opacityOffset = 1, minDistance$1 = 0;
    function getWarpDistance(pos1, pos2, canvasSize) {
        const { dx, dy } = engine.getDistances(pos1, pos2), absDiffs = { x: Math.abs(dx), y: Math.abs(dy) }, warpDistances = {
            x: Math.min(absDiffs.x, canvasSize.width - absDiffs.x),
            y: Math.min(absDiffs.y, canvasSize.height - absDiffs.y),
        };
        return Math.hypot(warpDistances.x, warpDistances.y);
    }
    class Linker extends pluginInteractivity.ParticlesInteractorBase {
        #maxDistance;
        #pluginManager;
        constructor(pluginManager, container) {
            super(container);
            this.#pluginManager = pluginManager;
            this.#maxDistance = 0;
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear() {
        }
        init() {
            this.container.particles.linksColor = undefined;
            this.container.particles.linksColors = new Map();
        }
        interact(p1) {
            if (!p1.options.links) {
                return;
            }
            p1.links = [];
            if (p1.linksDistance && p1.linksDistance > this.#maxDistance) {
                this.#maxDistance = p1.linksDistance;
            }
            const pos1 = p1.getPosition(), container = this.container, canvasSize = container.canvas.size;
            if (pos1.x < engine.originPoint.x || pos1.y < engine.originPoint.y || pos1.x > canvasSize.width || pos1.y > canvasSize.height) {
                return;
            }
            const linkOpt1 = p1.options.links, optOpacity = linkOpt1.opacity, optDistance = p1.retina.linksDistance ?? minDistance$1, warp = linkOpt1.warp, range = warp ? new CircleWarp(pos1.x, pos1.y, optDistance, canvasSize) : new engine.Circle(pos1.x, pos1.y, optDistance), query = container.particles.grid.query(range);
            for (const p2 of query) {
                const linkOpt2 = p2.options.links;
                if (p1 === p2 ||
                    !linkOpt2?.enable ||
                    linkOpt1.id !== linkOpt2.id ||
                    p2.spawning ||
                    p2.destroyed ||
                    !p2.links ||
                    p1.links.some(t => t.destination === p2) ||
                    p2.links.some(t => t.destination === p1)) {
                    continue;
                }
                const pos2 = p2.getPosition();
                if (pos2.x < engine.originPoint.x || pos2.y < engine.originPoint.y || pos2.x > canvasSize.width || pos2.y > canvasSize.height) {
                    continue;
                }
                const distDirect = engine.getDistances(pos1, pos2).distance, distWarp = warp && linkOpt2.warp ? getWarpDistance(pos1, pos2, canvasSize) : distDirect, distance = Math.min(distDirect, distWarp);
                if (distance > optDistance) {
                    continue;
                }
                const opacityLine = (opacityOffset - distance / optDistance) * optOpacity;
                this.#setColor(p1);
                p1.links.push({
                    destination: p2,
                    opacity: opacityLine,
                    color: this.#getLinkColor(p1, p2),
                    isWarped: distWarp < distDirect,
                });
            }
        }
        isEnabled(particle) {
            return !!particle.options.links?.enable;
        }
        loadParticlesOptions(options, ...sources) {
            options.links ??= new Links();
            for (const source of sources) {
                options.links.load(source?.links);
            }
        }
        reset() {
        }
        #getLinkColor(p1, p2) {
            const container = this.container, linksOptions = p1.options.links;
            if (!linksOptions) {
                return;
            }
            const linkColor = linksOptions.id !== undefined
                ? container.particles.linksColors.get(linksOptions.id)
                : container.particles.linksColor;
            return engine.getLinkColor(p1, p2, linkColor);
        }
        #setColor(p1) {
            if (!p1.options.links) {
                return;
            }
            const container = this.container, linksOptions = p1.options.links;
            let linkColor = linksOptions.id === undefined
                ? container.particles.linksColor
                : container.particles.linksColors.get(linksOptions.id);
            if (linkColor) {
                return;
            }
            linkColor = engine.getLinkRandomColor(this.#pluginManager, linksOptions.color, linksOptions.blink, linksOptions.consent);
            if (linksOptions.id === undefined) {
                container.particles.linksColor = linkColor;
            }
            else {
                container.particles.linksColors.set(linksOptions.id, linkColor);
            }
        }
    }

    class LinksPlugin {
        id = "links";
        #pluginManager;
        constructor(pluginManager) {
            this.#pluginManager = pluginManager;
        }
        async getPlugin(container) {
            const { LinkInstance } = await Promise.resolve().then(function () { return LinkInstance$1; });
            return new LinkInstance(this.#pluginManager, container);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }

    async function loadParticlesLinksInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            const pluginManager = e.pluginManager;
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            pluginManager.addPlugin(new LinksPlugin(pluginManager));
            pluginManager.addInteractor?.("particlesLinks", container => {
                return Promise.resolve(new Linker(pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadParticlesLinksInteraction = loadParticlesLinksInteraction;

    function getLinkKey(ids) {
        return [...ids].sort((a, b) => a - b).join("_");
    }
    function setLinkFrequency(particles, dictionary) {
        const key = getLinkKey(particles.map(t => t.id));
        let res = dictionary.get(key);
        if (res === undefined) {
            res = engine.getRandom();
            dictionary.set(key, res);
        }
        return res;
    }

    const minOpacity = 0, minWidth = 0, minDistance = 0, maxFrequency = 1, defaultFrequency = 0;
    class LinkInstance {
        #colorCache = new Map();
        #container;
        #freqs;
        #pluginManager;
        constructor(pluginManager, container) {
            this.#pluginManager = pluginManager;
            this.#container = container;
            this.#freqs = { links: new Map(), triangles: new Map() };
        }
        drawParticle(context, particle) {
            const { links, options } = particle;
            if (!links?.length || !options.links) {
                return;
            }
            const linkOpts = options.links, width = particle.retina.linksWidth ?? minWidth, pos1 = particle.getPosition(), twinkle = particle.options["twinkle"]?.links, trianglesEnabled = linkOpts.triangles.enable, p1Destinations = trianglesEnabled ? new Set(links.map(l => l.destination.id)) : null, originalAlpha = context.globalAlpha;
            let currentColorStyle = "", currentWidth = -1, currentAlpha = -1, pathOpen = false;
            const flushLines = () => {
                if (pathOpen) {
                    context.stroke();
                    pathOpen = false;
                }
            };
            for (const link of links) {
                if (linkOpts.frequency < maxFrequency &&
                    this.#getLinkFrequency(particle, link.destination) > linkOpts.frequency) {
                    continue;
                }
                const pos2 = link.destination.getPosition();
                if (trianglesEnabled && !link.isWarped && p1Destinations) {
                    flushLines();
                    this.#drawTriangles(options, particle, link, p1Destinations, pos1, pos2, context);
                }
                if (link.opacity <= minOpacity || width <= minWidth) {
                    continue;
                }
                if (!linkOpts.enable) {
                    continue;
                }
                let opacity = link.opacity, colorLine = link.color;
                const twinkleRgb = twinkle?.enable && engine.getRandom() < twinkle.frequency
                    ? engine.rangeColorToRgb(this.#pluginManager, twinkle.color)
                    : undefined;
                if (twinkle && twinkleRgb) {
                    colorLine = twinkleRgb;
                    opacity = engine.getRangeValue(twinkle.opacity);
                }
                if (!colorLine) {
                    const linkColor = linkOpts.id !== undefined
                        ? this.#container.particles.linksColors.get(linkOpts.id)
                        : this.#container.particles.linksColor;
                    colorLine = engine.getLinkColor(particle, link.destination, linkColor);
                }
                if (!colorLine) {
                    continue;
                }
                const colorStyle = this.#getCachedStyle(colorLine);
                if (colorStyle !== currentColorStyle || width !== currentWidth || opacity !== currentAlpha) {
                    flushLines();
                    context.strokeStyle = colorStyle;
                    context.lineWidth = width;
                    context.globalAlpha = opacity;
                    currentColorStyle = colorStyle;
                    currentWidth = width;
                    currentAlpha = opacity;
                    context.beginPath();
                    pathOpen = true;
                }
                if (link.isWarped) {
                    const canvasSize = this.#container.canvas.size, dx = pos2.x - pos1.x, dy = pos2.y - pos1.y;
                    let sx = engine.originPoint.x, sy = engine.originPoint.y;
                    if (Math.abs(dx) > canvasSize.width * engine.half) {
                        sx = dx > minDistance ? -canvasSize.width : canvasSize.width;
                    }
                    if (Math.abs(dy) > canvasSize.height * engine.half) {
                        sy = dy > minDistance ? -canvasSize.height : canvasSize.height;
                    }
                    context.moveTo(pos1.x, pos1.y);
                    context.lineTo(pos2.x + sx, pos2.y + sy);
                    context.moveTo(pos1.x - sx, pos1.y - sy);
                    context.lineTo(pos2.x, pos2.y);
                }
                else {
                    context.moveTo(pos1.x, pos1.y);
                    context.lineTo(pos2.x, pos2.y);
                }
            }
            flushLines();
            context.globalAlpha = originalAlpha;
        }
        init() {
            this.#freqs.links.clear();
            this.#freqs.triangles.clear();
            this.#colorCache.clear();
            return Promise.resolve();
        }
        particleCreated(particle) {
            particle.links = [];
            if (!particle.options.links) {
                return;
            }
            particle.linksDistance = particle.options.links.distance;
            particle.linksWidth = particle.options.links.width;
            const ratio = this.#container.retina.pixelRatio;
            particle.retina.linksDistance = particle.linksDistance * ratio;
            particle.retina.linksWidth = particle.linksWidth * ratio;
        }
        particleDestroyed(particle) {
            particle.links = [];
        }
        #drawTriangles(options, p1, link, p1Destinations, pos1, pos2, context) {
            const p2 = link.destination, triangleOptions = options.links?.triangles;
            if (!triangleOptions?.enable || !p2.options.links?.triangles.enable) {
                return;
            }
            const p2Links = p2.links;
            if (!p2Links?.length) {
                return;
            }
            for (const vertex of p2Links) {
                if (vertex.isWarped ||
                    this.#getLinkFrequency(p2, vertex.destination) > p2.options.links.frequency ||
                    !p1Destinations.has(vertex.destination.id)) {
                    continue;
                }
                const p3 = vertex.destination;
                if (this.#getTriangleFrequency(p1, p2, p3) > (options.links?.triangles.frequency ?? defaultFrequency)) {
                    continue;
                }
                const opacityTriangle = triangleOptions.opacity ?? (link.opacity + vertex.opacity) * engine.half, colorTriangle = engine.rangeColorToRgb(this.#pluginManager, triangleOptions.color) ?? link.color;
                if (!colorTriangle || opacityTriangle <= minOpacity) {
                    continue;
                }
                const pos3 = p3.getPosition();
                context.save();
                context.fillStyle = this.#getCachedStyle(colorTriangle);
                context.globalAlpha = opacityTriangle;
                context.beginPath();
                context.moveTo(pos1.x, pos1.y);
                context.lineTo(pos2.x, pos2.y);
                context.lineTo(pos3.x, pos3.y);
                context.closePath();
                context.fill();
                context.restore();
            }
        }
        #getCachedStyle(rgb) {
            const key = `${rgb.r},${rgb.g},${rgb.b}`;
            let style = this.#colorCache.get(key);
            if (!style) {
                style = engine.getStyleFromRgb(rgb, this.#container.hdr);
                this.#colorCache.set(key, style);
            }
            return style;
        }
        #getLinkFrequency(p1, p2) {
            return setLinkFrequency([p1, p2], this.#freqs.links);
        }
        #getTriangleFrequency(p1, p2, p3) {
            return setLinkFrequency([p1, p2, p3], this.#freqs.triangles);
        }
    }

    var LinkInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        LinkInstance: LinkInstance
    });

    exports.Links = Links;
    exports.LinksShadow = LinksShadow;
    exports.LinksTriangle = LinksTriangle;
    exports.loadParticlesLinksInteraction = loadParticlesLinksInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadParticlesLinksInteraction: (globalThis.__tsParticlesInternals.interactions.particlesLinks || {}).loadParticlesLinksInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
