// @tsparticles/shape-polygon/tsparticles.shape.polygon.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Shape v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.shapes = global.__tsParticlesInternals.shapes || {}, global.__tsParticlesInternals.shapes.polygon = global.__tsParticlesInternals.shapes.polygon || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    const polygonCache = new Map(), noOffset = 0;
    function getUnitPolygon(sides) {
        const cached = polygonCache.get(sides);
        if (cached) {
            return cached;
        }
        const step = engine.doublePI / sides, isOdd = !!(sides % engine.double), baseAngle = (-Math.PI + (isOdd ? noOffset : step)) * engine.half, verts = [];
        for (let i = 0; i < sides; i++) {
            const angle = baseAngle + i * step;
            verts[i] = {
                x: Math.cos(angle),
                y: Math.sin(angle),
            };
        }
        polygonCache.set(sides, verts);
        return verts;
    }
    function drawPolygon(data, side) {
        const { context, radius } = data, sides = side.count.numerator / side.count.denominator, verts = getUnitPolygon(sides);
        context.beginPath();
        for (let i = 0; i < verts.length; i++) {
            const vert = verts[i];
            if (!vert) {
                continue;
            }
            const x = vert.x * radius, y = vert.y * radius;
            if (i) {
                context.lineTo(x, y);
            }
            else {
                context.moveTo(x, y);
            }
        }
        context.closePath();
    }

    const defaultSides = 5;
    class PolygonDrawerBase {
        draw(data) {
            const { particle, radius } = data, side = this.getSidesData(particle, radius);
            drawPolygon(data, side);
        }
        getSidesCount(particle) {
            const polygon = particle.shapeData;
            return Math.round(engine.getRangeValue(polygon?.sides ?? defaultSides));
        }
    }

    const yFactor$1 = 2.66, sidesFactor$1 = 3;
    class PolygonDrawer extends PolygonDrawerBase {
        getSidesData(particle, radius) {
            const { sides } = particle;
            return {
                count: {
                    denominator: 1,
                    numerator: sides,
                },
                length: (radius * yFactor$1) / (sides / sidesFactor$1),
            };
        }
    }

    const sides = 3, yFactor = 2.66, sidesFactor = 3;
    class TriangleDrawer extends PolygonDrawerBase {
        getSidesCount() {
            return sides;
        }
        getSidesData(_particle, radius) {
            return {
                count: {
                    denominator: 1,
                    numerator: sides,
                },
                length: (radius * yFactor) / (sides / sidesFactor),
            };
        }
    }

    async function loadGenericPolygonShape(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register(e => {
            e.pluginManager.addShape(["polygon"], () => Promise.resolve(new PolygonDrawer()));
        });
    }
    async function loadTriangleShape(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register(e => {
            e.pluginManager.addShape(["triangle"], () => Promise.resolve(new TriangleDrawer()));
        });
    }
    async function loadPolygonShape(engine) {
        engine.checkVersion("4.1.3");
        await Promise.all([
            loadGenericPolygonShape(engine),
            loadTriangleShape(engine),
        ]);
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadGenericPolygonShape = loadGenericPolygonShape;
    globalObject.loadPolygonShape = loadPolygonShape;
    globalObject.loadTriangleShape = loadTriangleShape;

    exports.loadGenericPolygonShape = loadGenericPolygonShape;
    exports.loadPolygonShape = loadPolygonShape;
    exports.loadTriangleShape = loadTriangleShape;

}));
Object.assign(globalThis.window || globalThis, { loadGenericPolygonShape: (globalThis.__tsParticlesInternals.shapes.polygon || {}).loadGenericPolygonShape, loadPolygonShape: (globalThis.__tsParticlesInternals.shapes.polygon || {}).loadPolygonShape, loadTriangleShape: (globalThis.__tsParticlesInternals.shapes.polygon || {}).loadTriangleShape });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
