// @tsparticles/interaction-external-attract/browser/Attractor.js
import { ExternalInteractorBase, mouseMoveEvent, } from "@tsparticles/plugin-interactivity";
import { isInArray, millisecondsToSeconds } from "@tsparticles/engine";
import { clickAttract, hoverAttract } from "./Utils.js";
import { Attract } from "./Options/Classes/Attract.js";
const attractMode = "attract", minVelocityLengthSq = 0, minRestoreSpeed = 0.001, maxRestoreSpeed = 1, restoreEpsilon = 0.5;
export class Attractor extends ExternalInteractorBase {
    handleClickMode;
    #interactedThisFrame;
    #maxDistance;
    #pluginManager;
    #restoreData;
    constructor(pluginManager, container) {
        super(container);
        this.#pluginManager = pluginManager;
        this.#maxDistance = 0;
        this.#interactedThisFrame = new Set();
        this.#restoreData = new Map();
        container.attract ??= { particles: [] };
        this.handleClickMode = (mode, interactivityData) => {
            const options = this.container.actualOptions, attract = options.interactivity?.modes.attract;
            if (!attract || mode !== attractMode) {
                return;
            }
            container.attract ??= { particles: [] };
            container.attract.clicking = true;
            container.attract.count = 0;
            for (const particle of container.attract.particles) {
                if (!this.isEnabled(interactivityData, particle)) {
                    continue;
                }
                particle.velocity.setTo(particle.initialVelocity);
            }
            container.attract.particles = [];
            container.attract.finish = false;
            setTimeout(() => {
                if (container.destroyed) {
                    return;
                }
                container.attract ??= { particles: [] };
                container.attract.clicking = false;
            }, attract.duration * millisecondsToSeconds);
        };
    }
    get maxDistance() {
        return this.#maxDistance;
    }
    clear() {
    }
    init() {
        const container = this.container, attract = container.actualOptions.interactivity?.modes.attract;
        if (!attract) {
            return;
        }
        this.#maxDistance = attract.distance;
        container.retina.attractModeDistance = attract.distance * container.retina.pixelRatio;
    }
    interact(interactivityData) {
        this.#interactedThisFrame.clear();
        const container = this.container, options = container.actualOptions, mouseMoveStatus = interactivityData.status === mouseMoveEvent, events = options.interactivity?.events;
        if (!events) {
            return;
        }
        const { enable: hoverEnabled, mode: hoverMode } = events.onHover, { enable: clickEnabled, mode: clickMode } = events.onClick;
        if (mouseMoveStatus && hoverEnabled && isInArray(attractMode, hoverMode)) {
            hoverAttract(this.#pluginManager, this.container, interactivityData, p => this.isEnabled(interactivityData, p), p => {
                this.#trackInteractedParticle(p);
            });
        }
        else if (clickEnabled && isInArray(attractMode, clickMode)) {
            clickAttract(this.#pluginManager, this.container, interactivityData, p => this.isEnabled(interactivityData, p), p => {
                this.#trackInteractedParticle(p);
            });
        }
        this.#restoreParticles();
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
        if ((!mouse.position || !events?.onHover.enable) && (!mouse.clickPosition || !events?.onClick.enable)) {
            return false;
        }
        const hoverMode = events.onHover.mode, clickMode = events.onClick.mode;
        return isInArray(attractMode, hoverMode) || isInArray(attractMode, clickMode);
    }
    loadModeOptions(options, ...sources) {
        options.attract ??= new Attract();
        for (const source of sources) {
            options.attract.load(source?.attract);
        }
    }
    reset() {
    }
    #restoreParticles() {
        const restore = this.container.actualOptions.interactivity?.modes.attract?.restore;
        if (!restore?.enable || !this.#restoreData.size) {
            return;
        }
        const now = Date.now(), restoreDelay = restore.delay * millisecondsToSeconds, restoreSpeed = Math.max(minRestoreSpeed, Math.min(maxRestoreSpeed, restore.speed));
        for (const [particle, restoreData] of this.#restoreData) {
            if (this.#interactedThisFrame.has(particle)) {
                continue;
            }
            if (particle.destroyed) {
                this.#restoreData.delete(particle);
                continue;
            }
            const target = restoreData.target;
            if (now - restoreData.lastInteractionTime < restoreDelay) {
                continue;
            }
            let dx = target.x - particle.position.x, dy = target.y - particle.position.y, dz = target.z - particle.position.z;
            if (restore.follow && particle.options.move.enable) {
                const { x: vx, y: vy, z: vz } = particle.velocity, velocityLengthSq = vx * vx + vy * vy + vz * vz;
                if (velocityLengthSq > minVelocityLengthSq) {
                    const parallelScale = (dx * vx + dy * vy + dz * vz) / velocityLengthSq;
                    dx -= vx * parallelScale;
                    dy -= vy * parallelScale;
                    dz -= vz * parallelScale;
                }
            }
            particle.position.x += dx * restoreSpeed;
            particle.position.y += dy * restoreSpeed;
            particle.position.z += dz * restoreSpeed;
            if (Math.abs(dx) <= restoreEpsilon && Math.abs(dy) <= restoreEpsilon) {
                particle.position.x = target.x;
                particle.position.y = target.y;
                particle.position.z = target.z;
                this.#restoreData.delete(particle);
                continue;
            }
        }
    }
    #trackInteractedParticle(particle) {
        this.#interactedThisFrame.add(particle);
        const restore = this.container.actualOptions.interactivity?.modes.attract?.restore;
        if (!restore?.enable) {
            return;
        }
        const now = Date.now();
        let restoreData = this.#restoreData.get(particle);
        if (!restoreData) {
            restoreData = {
                target: particle.position.copy(),
                lastInteractionTime: now,
            };
            this.#restoreData.set(particle, restoreData);
        }
        restoreData.lastInteractionTime = now;
    }
}
