// @tsparticles/plugin-interactivity/browser/InteractionManager.js
import { safeIntersectionObserver, } from "@tsparticles/engine";
import { clickEvent, touchCancelEvent, touchEndEvent, touchMoveEvent, touchStartEvent, } from "./InteractivityConstants.js";
import { InteractivityEventListeners } from "./InteractivityEventListeners.js";
import { InteractorType } from "./Enums/InteractorType.js";
const clickRadius = 1, touchEndLengthOffset = 1, minCoordinate = 0;
export class InteractionManager {
    interactivityData;
    #clickHandlers;
    #container;
    #eventListeners;
    #externalInteractors;
    #interactors;
    #intersectionObserver;
    #particleInteractors;
    #pluginManager;
    constructor(pluginManager, container) {
        this.#container = container;
        this.#pluginManager = pluginManager;
        this.#interactors = [];
        this.#externalInteractors = [];
        this.#particleInteractors = [];
        this.#clickHandlers = new Map();
        this.#eventListeners = new InteractivityEventListeners(container, this);
        this.interactivityData = {
            mouse: {
                clicking: false,
                inside: false,
            },
        };
        this.#intersectionObserver = safeIntersectionObserver(entries => {
            this.#intersectionManager(entries);
        });
    }
    addClickHandler(callback) {
        const container = this.#container, interactivityData = this.interactivityData;
        if (container.destroyed) {
            return;
        }
        const el = interactivityData.element;
        if (!el) {
            return;
        }
        const clickOrTouchHandler = (e, pos, radius) => {
            if (container.destroyed) {
                return;
            }
            const pxRatio = container.retina.pixelRatio, posRetina = {
                x: pos.x * pxRatio,
                y: pos.y * pxRatio,
            }, particles = container.particles.grid.queryCircle(posRetina, radius * pxRatio);
            callback(e, particles);
        }, clickHandler = (e) => {
            if (container.destroyed) {
                return;
            }
            const mouseEvent = e, pos = {
                x: mouseEvent.offsetX,
                y: mouseEvent.offsetY,
            };
            clickOrTouchHandler(e, pos, clickRadius);
        }, touchStartHandler = () => {
            if (container.destroyed) {
                return;
            }
            touched = true;
            touchMoved = false;
        }, touchMoveHandler = () => {
            if (container.destroyed) {
                return;
            }
            touchMoved = true;
        }, touchEndHandler = (e) => {
            if (container.destroyed) {
                return;
            }
            if (touched && !touchMoved) {
                const touchEvent = e, lastTouch = touchEvent.touches[touchEvent.touches.length - touchEndLengthOffset];
                if (!lastTouch) {
                    return;
                }
                const element = container.canvas.domElement, canvasRect = element ? element.getBoundingClientRect() : undefined, pos = {
                    x: lastTouch.clientX - (canvasRect ? canvasRect.left : minCoordinate),
                    y: lastTouch.clientY - (canvasRect ? canvasRect.top : minCoordinate),
                };
                clickOrTouchHandler(e, pos, Math.max(lastTouch.radiusX, lastTouch.radiusY));
            }
            touched = false;
            touchMoved = false;
        }, touchCancelHandler = () => {
            if (container.destroyed) {
                return;
            }
            touched = false;
            touchMoved = false;
        };
        let touched = false, touchMoved = false;
        this.#clickHandlers.set(clickEvent, clickHandler);
        this.#clickHandlers.set(touchStartEvent, touchStartHandler);
        this.#clickHandlers.set(touchMoveEvent, touchMoveHandler);
        this.#clickHandlers.set(touchEndEvent, touchEndHandler);
        this.#clickHandlers.set(touchCancelEvent, touchCancelHandler);
        for (const [key, handler] of this.#clickHandlers) {
            el.addEventListener(key, handler);
        }
    }
    addListeners() {
        this.#eventListeners.addListeners();
    }
    clearClickHandlers() {
        const container = this.#container, interactivityData = this.interactivityData;
        if (container.destroyed) {
            return;
        }
        for (const [key, handler] of this.#clickHandlers) {
            interactivityData.element?.removeEventListener(key, handler);
        }
        this.#clickHandlers.clear();
    }
    externalInteract(delta) {
        for (const interactor of this.#externalInteractors) {
            const interactivityData = this.interactivityData;
            if (interactor.isEnabled(interactivityData)) {
                interactor.interact(interactivityData, delta);
            }
        }
    }
    handleClickMode(mode) {
        if (this.#container.destroyed) {
            return;
        }
        const interactivityData = this.interactivityData;
        for (const interactor of this.#externalInteractors) {
            interactor.handleClickMode?.(mode, interactivityData);
        }
    }
    init() {
        this.#eventListeners.init();
        for (const interactor of this.#interactors) {
            switch (interactor.type) {
                case InteractorType.external:
                    this.#externalInteractors.push(interactor);
                    break;
                case InteractorType.particles:
                    this.#particleInteractors.push(interactor);
                    break;
            }
            interactor.init();
        }
    }
    async initInteractors() {
        const interactors = await this.#pluginManager.getInteractors?.(this.#container, true);
        if (!interactors) {
            return;
        }
        this.#interactors = interactors;
        this.#externalInteractors = [];
        this.#particleInteractors = [];
    }
    particlesInteract(particle, delta) {
        const interactivityData = this.interactivityData;
        for (const interactor of this.#externalInteractors) {
            interactor.clear(particle, delta);
        }
        for (const interactor of this.#particleInteractors) {
            if (interactor.isEnabled(particle, interactivityData)) {
                interactor.interact(particle, interactivityData, delta);
            }
        }
    }
    removeListeners() {
        this.#eventListeners.removeListeners();
    }
    reset(particle) {
        const interactivityData = this.interactivityData;
        for (const interactor of this.#externalInteractors) {
            if (interactor.isEnabled(interactivityData)) {
                interactor.reset(interactivityData, particle);
            }
        }
        for (const interactor of this.#particleInteractors) {
            if (interactor.isEnabled(particle, interactivityData)) {
                interactor.reset(interactivityData, particle);
            }
        }
    }
    startObserving() {
        const interactivityData = this.interactivityData;
        if (interactivityData.element instanceof HTMLElement && this.#intersectionObserver) {
            this.#intersectionObserver.observe(interactivityData.element);
        }
    }
    stopObserving() {
        const interactivityData = this.interactivityData;
        if (interactivityData.element instanceof HTMLElement && this.#intersectionObserver) {
            this.#intersectionObserver.unobserve(interactivityData.element);
        }
    }
    updateMaxDistance() {
        let maxTotalDistance = 0;
        for (const interactor of this.#interactors) {
            if (interactor.maxDistance > maxTotalDistance) {
                maxTotalDistance = interactor.maxDistance;
            }
        }
        const container = this.#container;
        container.particles.grid.setCellSize(maxTotalDistance * container.retina.pixelRatio);
    }
    #intersectionManager = entries => {
        const container = this.#container;
        if (container.destroyed || !container.actualOptions.pauseOnOutsideViewport) {
            return;
        }
        for (const entry of entries) {
            if (entry.target !== this.interactivityData.element) {
                continue;
            }
            if (entry.isIntersecting) {
                container.play();
            }
            else {
                container.pause();
            }
        }
    };
}
