// @tsparticles/interaction-particles-links/esm/Linker.js
import { Circle, getDistances, getLinkColor, getLinkRandomColor, originPoint, } from "@tsparticles/engine";
import { CircleWarp } from "./CircleWarp.js";
import { Links } from "./Options/Classes/Links.js";
import { ParticlesInteractorBase } from "@tsparticles/plugin-interactivity";
const opacityOffset = 1, minDistance = 0;
function getWarpDistance(pos1, pos2, canvasSize) {
    const { dx, dy } = getDistances(pos1, pos2), absDiffs = { x: Math.abs(dx), y: Math.abs(dy) }, warpDistances = {
        x: Math.min(absDiffs.x, canvasSize.width - absDiffs.x),
        y: Math.min(absDiffs.y, canvasSize.height - absDiffs.y),
    };
    return Math.hypot(warpDistances.x, warpDistances.y);
}
export class Linker extends ParticlesInteractorBase {
    #maxDistance;
    #pluginManager;
    constructor(pluginManager, container) {
        super(container);
        this.#pluginManager = pluginManager;
        this.#maxDistance = 0;
    }
    get maxDistance() {
        return this.#maxDistance;
    }
    clear() {
    }
    init() {
        this.container.particles.linksColor = undefined;
        this.container.particles.linksColors = new Map();
    }
    interact(p1) {
        if (!p1.options.links) {
            return;
        }
        p1.links = [];
        if (p1.linksDistance && p1.linksDistance > this.#maxDistance) {
            this.#maxDistance = p1.linksDistance;
        }
        const pos1 = p1.getPosition(), container = this.container, canvasSize = container.canvas.size;
        if (pos1.x < originPoint.x || pos1.y < originPoint.y || pos1.x > canvasSize.width || pos1.y > canvasSize.height) {
            return;
        }
        const linkOpt1 = p1.options.links, optOpacity = linkOpt1.opacity, optDistance = p1.retina.linksDistance ?? minDistance, warp = linkOpt1.warp, range = warp ? new CircleWarp(pos1.x, pos1.y, optDistance, canvasSize) : new Circle(pos1.x, pos1.y, optDistance), query = container.particles.grid.query(range);
        for (const p2 of query) {
            const linkOpt2 = p2.options.links;
            if (p1 === p2 ||
                !linkOpt2?.enable ||
                linkOpt1.id !== linkOpt2.id ||
                p2.spawning ||
                p2.destroyed ||
                !p2.links ||
                p1.links.some(t => t.destination === p2) ||
                p2.links.some(t => t.destination === p1)) {
                continue;
            }
            const pos2 = p2.getPosition();
            if (pos2.x < originPoint.x || pos2.y < originPoint.y || pos2.x > canvasSize.width || pos2.y > canvasSize.height) {
                continue;
            }
            const distDirect = getDistances(pos1, pos2).distance, distWarp = warp && linkOpt2.warp ? getWarpDistance(pos1, pos2, canvasSize) : distDirect, distance = Math.min(distDirect, distWarp);
            if (distance > optDistance) {
                continue;
            }
            const opacityLine = (opacityOffset - distance / optDistance) * optOpacity;
            this.#setColor(p1);
            p1.links.push({
                destination: p2,
                opacity: opacityLine,
                color: this.#getLinkColor(p1, p2),
                isWarped: distWarp < distDirect,
            });
        }
    }
    isEnabled(particle) {
        return !!particle.options.links?.enable;
    }
    loadParticlesOptions(options, ...sources) {
        options.links ??= new Links();
        for (const source of sources) {
            options.links.load(source?.links);
        }
    }
    reset() {
    }
    #getLinkColor(p1, p2) {
        const container = this.container, linksOptions = p1.options.links;
        if (!linksOptions) {
            return;
        }
        const linkColor = linksOptions.id !== undefined
            ? container.particles.linksColors.get(linksOptions.id)
            : container.particles.linksColor;
        return getLinkColor(p1, p2, linkColor);
    }
    #setColor(p1) {
        if (!p1.options.links) {
            return;
        }
        const container = this.container, linksOptions = p1.options.links;
        let linkColor = linksOptions.id === undefined
            ? container.particles.linksColor
            : container.particles.linksColors.get(linksOptions.id);
        if (linkColor) {
            return;
        }
        linkColor = getLinkRandomColor(this.#pluginManager, linksOptions.color, linksOptions.blink, linksOptions.consent);
        if (linksOptions.id === undefined) {
            container.particles.linksColor = linkColor;
        }
        else {
            container.particles.linksColors.set(linksOptions.id, linkColor);
        }
    }
}
