// @tsparticles/interaction-external-destroy/cjs/Utils.js
import { Circle, Rectangle, double, safeDocument } from "@tsparticles/engine";
import { DivType, divModeExecute } from "@tsparticles/plugin-interactivity";
const half = 0.5, toleranceFactor = 10, minRadius = 0;
function processDestroy(container, area) {
    const query = container.particles.grid.query(area);
    for (const particle of query) {
        particle.destroy();
    }
}
function singleSelectorDestroy(container, selector, div) {
    const query = safeDocument().querySelectorAll(selector);
    if (!query.length) {
        return;
    }
    query.forEach(item => {
        const elem = item, pxRatio = container.retina.pixelRatio, pos = {
            x: (elem.offsetLeft + elem.offsetWidth * half) * pxRatio,
            y: (elem.offsetTop + elem.offsetHeight * half) * pxRatio,
        }, radius = elem.offsetWidth * half * pxRatio, tolerance = toleranceFactor * pxRatio, area = div.type === DivType.circle
            ? new Circle(pos.x, pos.y, radius + tolerance)
            : new Rectangle(elem.offsetLeft * pxRatio - tolerance, elem.offsetTop * pxRatio - tolerance, elem.offsetWidth * pxRatio + tolerance * double, elem.offsetHeight * pxRatio + tolerance * double);
        processDestroy(container, area);
    });
}
export function divDestroy(container, divs, destroyMode) {
    divModeExecute(destroyMode, divs, (selector, div) => {
        singleSelectorDestroy(container, selector, div);
    });
}
export function mouseDestroy(container, interactivityData) {
    const pxRatio = container.retina.pixelRatio, tolerance = toleranceFactor * pxRatio, mousePos = interactivityData.mouse.position, radius = container.retina.destroyModeDistance;
    if (!radius || radius < minRadius || !mousePos) {
        return;
    }
    processDestroy(container, new Circle(mousePos.x, mousePos.y, radius + tolerance));
}
