// @tsparticles/plugin-move/tsparticles.plugin.move.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.move = global.__tsParticlesInternals.plugins.move || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class MovePlugin {
        id = "move";
        #pluginManager;
        constructor(pluginManager) {
            this.#pluginManager = pluginManager;
        }
        async getPlugin(container) {
            const { MovePluginInstance } = await Promise.resolve().then(function () { return MovePluginInstance$1; });
            return new MovePluginInstance(this.#pluginManager, container);
        }
        loadOptions() {
        }
        needsPlugin() {
            return true;
        }
    }

    async function loadMovePlugin(engine$1) {
        engine$1.checkVersion("4.1.3");
        await engine$1.pluginManager.register(e => {
            const moveEngine = e, movePluginManager = moveEngine.pluginManager;
            movePluginManager.initializers.pathGenerators ??= new Map();
            movePluginManager.pathGenerators ??= new Map();
            movePluginManager.addPathGenerator = (name, generator) => {
                movePluginManager.initializers.pathGenerators ??= new Map();
                movePluginManager.initializers.pathGenerators.set(name, generator);
            };
            movePluginManager.getPathGenerators = async (container, force = false) => {
                movePluginManager.initializers.pathGenerators ??= new Map();
                movePluginManager.pathGenerators ??= new Map();
                return engine.getItemMapFromInitializer(container, movePluginManager.pathGenerators, movePluginManager.initializers.pathGenerators, force);
            };
            e.pluginManager.addPlugin(new MovePlugin(e.pluginManager));
        });
    }
    function ensureBaseMoverLoaded(e) {
        if (!e.pluginManager.addPathGenerator) {
            throw new Error("tsParticles Base Mover is not loaded");
        }
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadMovePlugin = loadMovePlugin;

    const minVelocity = 0, identity = 1, moveSpeedFactor = 60, minSpinRadius = 0, spinFactor = 0.01, defaultPathDelay = 0, noDecay = 1;
    function applyDistance(particle) {
        const initialPosition = particle.initialPosition, { dx, dy } = engine.getDistances(initialPosition, particle.position), dxFixed = Math.abs(dx), dyFixed = Math.abs(dy), { maxDistance } = particle.retina, hDistance = maxDistance.horizontal, vDistance = maxDistance.vertical;
        if (!hDistance && !vDistance) {
            return;
        }
        const hasHDistance = (hDistance && dxFixed >= hDistance) ?? false, hasVDistance = (vDistance && dyFixed >= vDistance) ?? false;
        if ((hasHDistance || hasVDistance) && !particle.misplaced) {
            particle.misplaced = (!!hDistance && dxFixed > hDistance) || (!!vDistance && dyFixed > vDistance);
            if (hDistance) {
                particle.velocity.x = particle.velocity.y * engine.half - particle.velocity.x;
            }
            if (vDistance) {
                particle.velocity.y = particle.velocity.x * engine.half - particle.velocity.y;
            }
        }
        else if ((!hDistance || dxFixed < hDistance) && (!vDistance || dyFixed < vDistance) && particle.misplaced) {
            particle.misplaced = false;
        }
        else if (particle.misplaced) {
            const pos = particle.position, vel = particle.velocity;
            if (hDistance &&
                ((pos.x < initialPosition.x && vel.x < minVelocity) || (pos.x > initialPosition.x && vel.x > minVelocity))) {
                vel.x *= -engine.getRandom();
            }
            if (vDistance &&
                ((pos.y < initialPosition.y && vel.y < minVelocity) || (pos.y > initialPosition.y && vel.y > minVelocity))) {
                vel.y *= -engine.getRandom();
            }
        }
    }
    function move(particle, moveOptions, moveSpeed, maxSpeed, moveDrift, reduceFactor, delta) {
        applyPath(particle, delta);
        const gravityOptions = particle.gravity, gravityFactor = gravityOptions?.enable && gravityOptions.inverse ? -identity : identity;
        if (moveDrift && moveSpeed) {
            particle.velocity.x += (moveDrift * delta.factor) / (moveSpeedFactor * moveSpeed);
        }
        if (gravityOptions?.enable && moveSpeed) {
            particle.velocity.y +=
                (gravityFactor * (gravityOptions.acceleration * delta.factor)) / (moveSpeedFactor * moveSpeed);
        }
        const decay = particle.moveDecay;
        particle.velocity.multTo(decay ?? noDecay);
        const velocity = particle.velocity.mult(moveSpeed);
        if (gravityOptions?.enable &&
            maxSpeed > minVelocity &&
            ((!gravityOptions.inverse && velocity.y >= minVelocity && velocity.y >= maxSpeed) ||
                (gravityOptions.inverse && velocity.y <= minVelocity && velocity.y <= -maxSpeed))) {
            velocity.y = gravityFactor * maxSpeed;
            if (moveSpeed) {
                particle.velocity.y = velocity.y / moveSpeed;
            }
        }
        const zIndexOptions = particle.options.zIndex, zVelocityFactor = (identity - particle.zIndexFactor) ** zIndexOptions.velocityRate;
        velocity.multTo(zVelocityFactor);
        velocity.multTo(reduceFactor);
        const { position } = particle;
        position.addTo(velocity);
        if (moveOptions.vibrate) {
            position.x += Math.sin(position.x * Math.cos(position.y)) * reduceFactor;
            position.y += Math.cos(position.y * Math.sin(position.x)) * reduceFactor;
        }
    }
    function spin(container, particle, moveSpeed, reduceFactor) {
        if (!particle.spin) {
            return;
        }
        const spinClockwise = particle.spin.direction === engine.RotateDirection.clockwise, updateFunc = {
            x: spinClockwise ? Math.cos : Math.sin,
            y: spinClockwise ? Math.sin : Math.cos,
        };
        particle.position.x =
            particle.spin.center.x + particle.spin.radius * updateFunc.x(particle.spin.angle) * reduceFactor;
        particle.position.y =
            particle.spin.center.y + particle.spin.radius * updateFunc.y(particle.spin.angle) * reduceFactor;
        particle.spin.radius += particle.spin.acceleration * reduceFactor;
        const maxCanvasSize = Math.max(container.canvas.size.width, container.canvas.size.height), halfMaxSize = maxCanvasSize * engine.half;
        if (particle.spin.radius > halfMaxSize) {
            particle.spin.radius = halfMaxSize;
            particle.spin.acceleration *= -identity;
        }
        else if (particle.spin.radius < minSpinRadius) {
            particle.spin.radius = minSpinRadius;
            particle.spin.acceleration *= -identity;
        }
        particle.spin.angle += moveSpeed * spinFactor * (identity - particle.spin.radius / maxCanvasSize);
    }
    function applyPath(particle, delta) {
        const particlesOptions = particle.options, pathOptions = particlesOptions.move.path, pathEnabled = pathOptions.enable;
        if (!pathEnabled) {
            return;
        }
        const pathDelay = particle.pathDelay ?? defaultPathDelay;
        if (particle.lastPathTime <= pathDelay) {
            particle.lastPathTime += delta.value;
            return;
        }
        const path = particle.pathGenerator?.generate(particle, delta);
        if (path) {
            particle.velocity.addTo(path);
        }
        if (pathOptions.clamp) {
            particle.velocity.x = engine.clamp(particle.velocity.x, -identity, identity);
            particle.velocity.y = engine.clamp(particle.velocity.y, -identity, identity);
        }
        particle.lastPathTime -= pathDelay;
    }
    function getProximitySpeedFactor(particle) {
        return particle.slow.inRange ? particle.slow.factor : identity;
    }
    function initSpin(container, particle) {
        const options = particle.options, spinOptions = options.move.spin;
        if (!spinOptions.enable) {
            return;
        }
        const spinPos = spinOptions.position ?? { x: 50, y: 50 }, spinFactor = 0.01, spinCenter = {
            x: spinPos.x * spinFactor * container.canvas.size.width,
            y: spinPos.y * spinFactor * container.canvas.size.height,
        }, pos = particle.getPosition(), distance = engine.getDistance(pos, spinCenter), spinAcceleration = engine.getRangeValue(spinOptions.acceleration);
        particle.retina.spinAcceleration = spinAcceleration * container.retina.pixelRatio;
        particle.spin = {
            center: spinCenter,
            direction: particle.velocity.x >= minVelocity ? engine.RotateDirection.clockwise : engine.RotateDirection.counterClockwise,
            angle: engine.getRandom() * engine.doublePI,
            radius: distance,
            acceleration: particle.retina.spinAcceleration,
        };
    }

    const defaultSizeFactor = 1, defaultDeltaFactor = 1;
    class MovePluginInstance {
        availablePathGenerators;
        pathGenerators;
        #container;
        #pluginManager;
        constructor(pluginManager, container) {
            this.#pluginManager = pluginManager;
            this.#container = container;
            this.availablePathGenerators = new Map();
            this.pathGenerators = new Map();
        }
        destroy() {
            this.availablePathGenerators = new Map();
            this.pathGenerators = new Map();
        }
        isEnabled(particle) {
            return !particle.destroyed && particle.options.move.enable;
        }
        particleCreated(particle) {
            const options = particle.options, moveOptions = options.move, gravityOptions = moveOptions.gravity, pathOptions = moveOptions.path;
            particle.moveDecay = engine.decayOffset - engine.getRangeValue(moveOptions.decay);
            particle.pathDelay = engine.getRangeValue(pathOptions.delay.value) * engine.millisecondsToSeconds;
            if (pathOptions.generator) {
                let pathGenerator = this.pathGenerators.get(pathOptions.generator);
                if (!pathGenerator) {
                    pathGenerator = this.availablePathGenerators.get(pathOptions.generator);
                    if (pathGenerator) {
                        this.pathGenerators.set(pathOptions.generator, pathGenerator);
                        pathGenerator.init();
                    }
                }
                particle.pathGenerator = pathGenerator;
            }
            particle.gravity = {
                enable: gravityOptions.enable,
                acceleration: engine.getRangeValue(gravityOptions.acceleration),
                inverse: gravityOptions.inverse,
            };
            initSpin(this.#container, particle);
        }
        particleDestroyed(particle) {
            const pathGenerator = particle.pathGenerator;
            pathGenerator?.reset(particle);
        }
        particleUpdate(particle, delta) {
            const particleOptions = particle.options, moveOptions = particleOptions.move;
            if (!moveOptions.enable) {
                return;
            }
            const container = this.#container, slowFactor = getProximitySpeedFactor(particle), reduceFactor = container.retina.reduceFactor, baseSpeed = particle.retina.moveSpeed, moveDrift = particle.retina.moveDrift, maxSize = particle.size.max, sizeFactor = moveOptions.size ? particle.getRadius() / maxSize : defaultSizeFactor, deltaFactor = delta.factor || defaultDeltaFactor, moveSpeed = baseSpeed * sizeFactor * slowFactor * deltaFactor * engine.half, maxSpeed = particle.retina.maxSpeed;
            if (moveOptions.spin.enable) {
                spin(container, particle, moveSpeed, reduceFactor);
            }
            else {
                move(particle, moveOptions, moveSpeed, maxSpeed, moveDrift, reduceFactor, delta);
            }
            applyDistance(particle);
        }
        preInit() {
            return this.#init();
        }
        redrawInit() {
            return this.#init();
        }
        update() {
            for (const pathGenerator of this.pathGenerators.values()) {
                pathGenerator.update();
            }
        }
        async #init() {
            const availablePathGenerators = await this.#pluginManager.getPathGenerators?.(this.#container, true);
            if (!availablePathGenerators) {
                return;
            }
            this.availablePathGenerators = availablePathGenerators;
            this.pathGenerators = new Map();
            for (const pathGenerator of this.pathGenerators.values()) {
                pathGenerator.init();
            }
        }
    }

    var MovePluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        MovePluginInstance: MovePluginInstance
    });

    exports.ensureBaseMoverLoaded = ensureBaseMoverLoaded;
    exports.loadMovePlugin = loadMovePlugin;

}));
Object.assign(globalThis.window || globalThis, { loadMovePlugin: (globalThis.__tsParticlesInternals.plugins.move || {}).loadMovePlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
