// @tsparticles/interaction-external-attract/browser/Options/Classes/Attract.js
import { EasingType, isNull } from "@tsparticles/engine";
export class Attract {
    distance;
    duration;
    easing;
    factor;
    maxSpeed;
    restore;
    speed;
    constructor() {
        this.distance = 200;
        this.duration = 0.4;
        this.easing = EasingType.easeOutQuad;
        this.factor = 1;
        this.maxSpeed = 50;
        this.speed = 1;
        this.restore = {
            enable: false,
            delay: 0,
            speed: 0.08,
            follow: true,
        };
    }
    load(data) {
        if (isNull(data)) {
            return;
        }
        if (data.distance !== undefined) {
            this.distance = data.distance;
        }
        if (data.duration !== undefined) {
            this.duration = data.duration;
        }
        if (data.easing !== undefined) {
            this.easing = data.easing;
        }
        if (data.factor !== undefined) {
            this.factor = data.factor;
        }
        if (data.maxSpeed !== undefined) {
            this.maxSpeed = data.maxSpeed;
        }
        if (data.speed !== undefined) {
            this.speed = data.speed;
        }
        if (data.restore !== undefined) {
            this.restore.enable = data.restore.enable ?? this.restore.enable;
            this.restore.delay = data.restore.delay ?? this.restore.delay;
            this.restore.speed = data.restore.speed ?? this.restore.speed;
            this.restore.follow = data.restore.follow ?? this.restore.follow;
        }
    }
}
