// @tsparticles/updater-opacity/tsparticles.updater.opacity.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.opacity = global.__tsParticlesInternals.updaters.opacity || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class OpacityAnimation extends engine.RangedAnimationOptions {
        destroy;
        constructor() {
            super();
            this.destroy = engine.DestroyType.none;
            this.speed = 2;
        }
        load(data) {
            super.load(data);
            if (engine.isNull(data)) {
                return;
            }
            if (data.destroy !== undefined) {
                this.destroy = data.destroy;
            }
        }
    }

    class Opacity extends engine.RangedAnimationValueWithRandom {
        animation;
        constructor() {
            super();
            this.animation = new OpacityAnimation();
            this.value = 1;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            super.load(data);
            const animation = data.animation;
            if (animation !== undefined) {
                this.animation.load(animation);
            }
        }
    }

    class OpacityUpdater {
        #container;
        constructor(container) {
            this.#container = container;
        }
        init(particle) {
            const opacityOptions = particle.options.opacity, pxRatio = 1;
            if (!opacityOptions) {
                return;
            }
            particle.opacity = engine.initParticleNumericAnimationValue(opacityOptions, pxRatio);
            const opacityAnimation = opacityOptions.animation;
            if (opacityAnimation.enable) {
                particle.opacity.velocity =
                    (engine.getRangeValue(opacityAnimation.speed) / engine.percentDenominator) * this.#container.retina.reduceFactor;
                if (!opacityAnimation.sync) {
                    particle.opacity.velocity *= engine.getRandom();
                }
            }
        }
        isEnabled(particle) {
            const none = 0;
            return (!particle.destroyed &&
                !particle.spawning &&
                !!particle.opacity &&
                particle.opacity.enable &&
                ((particle.opacity.maxLoops ?? none) <= none ||
                    ((particle.opacity.maxLoops ?? none) > none &&
                        (particle.opacity.loops ?? none) < (particle.opacity.maxLoops ?? none))));
        }
        loadOptions(options, ...sources) {
            options.opacity ??= new Opacity();
            for (const source of sources) {
                options.opacity.load(source?.opacity);
            }
        }
        reset(particle) {
            if (!particle.opacity) {
                return;
            }
            particle.opacity.time = 0;
            particle.opacity.loops = 0;
        }
        update(particle, delta) {
            if (!this.isEnabled(particle) || !particle.opacity || !particle.options.opacity) {
                return;
            }
            engine.updateAnimation(particle, particle.opacity, true, particle.options.opacity.animation.destroy, delta);
        }
    }

    async function loadOpacityUpdater(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("opacity", container => {
                return Promise.resolve(new OpacityUpdater(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadOpacityUpdater = loadOpacityUpdater;

    exports.loadOpacityUpdater = loadOpacityUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadOpacityUpdater: (globalThis.__tsParticlesInternals.updaters.opacity || {}).loadOpacityUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
