// @tsparticles/interaction-external-attract/tsparticles.interaction.external.attract.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalAttract = global.__tsParticlesInternals.interactions.externalAttract || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    const minFactor = 1, minRadius = 0, updateVector = engine.Vector.origin;
    function processAttract(pluginManager, container, position, attractRadius, area, queryCb, onAttractParticle) {
        const attractOptions = container.actualOptions.interactivity?.modes.attract;
        if (!attractOptions) {
            return;
        }
        const query = container.particles.grid.query(area, queryCb);
        for (const particle of query) {
            const { dx, dy, distance } = engine.getDistances(particle.position, position), velocity = attractOptions.speed * attractOptions.factor, attractFactor = engine.clamp(pluginManager.getEasing(attractOptions.easing)(engine.identity - distance / attractRadius) * velocity, minFactor, attractOptions.maxSpeed);
            updateVector.x = !distance ? velocity : (dx / distance) * attractFactor;
            updateVector.y = !distance ? velocity : (dy / distance) * attractFactor;
            onAttractParticle?.(particle);
            particle.position.subFrom(updateVector);
        }
    }
    function clickAttract(pluginManager, container, interactivityData, enabledCb, onAttractParticle) {
        container.attract ??= { particles: [] };
        const { attract } = container;
        if (!attract.finish) {
            attract.count ??= 0;
            attract.count++;
            if (attract.count === container.particles.count) {
                attract.finish = true;
            }
        }
        if (attract.clicking) {
            const mousePos = interactivityData.mouse.clickPosition, attractRadius = container.retina.attractModeDistance;
            if (!attractRadius || attractRadius < minRadius || !mousePos) {
                return;
            }
            processAttract(pluginManager, container, mousePos, attractRadius, new engine.Circle(mousePos.x, mousePos.y, attractRadius), (p) => enabledCb(p), onAttractParticle);
        }
        else if (attract.clicking === false) {
            attract.particles = [];
        }
    }
    function hoverAttract(pluginManager, container, interactivityData, enabledCb, onAttractParticle) {
        const mousePos = interactivityData.mouse.position, attractRadius = container.retina.attractModeDistance;
        if (!attractRadius || attractRadius < minRadius || !mousePos) {
            return;
        }
        processAttract(pluginManager, container, mousePos, attractRadius, new engine.Circle(mousePos.x, mousePos.y, attractRadius), (p) => enabledCb(p), onAttractParticle);
    }

    class Attract {
        distance;
        duration;
        easing;
        factor;
        maxSpeed;
        restore;
        speed;
        constructor() {
            this.distance = 200;
            this.duration = 0.4;
            this.easing = engine.EasingType.easeOutQuad;
            this.factor = 1;
            this.maxSpeed = 50;
            this.speed = 1;
            this.restore = {
                enable: false,
                delay: 0,
                speed: 0.08,
                follow: true,
            };
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.distance !== undefined) {
                this.distance = data.distance;
            }
            if (data.duration !== undefined) {
                this.duration = data.duration;
            }
            if (data.easing !== undefined) {
                this.easing = data.easing;
            }
            if (data.factor !== undefined) {
                this.factor = data.factor;
            }
            if (data.maxSpeed !== undefined) {
                this.maxSpeed = data.maxSpeed;
            }
            if (data.speed !== undefined) {
                this.speed = data.speed;
            }
            if (data.restore !== undefined) {
                this.restore.enable = data.restore.enable ?? this.restore.enable;
                this.restore.delay = data.restore.delay ?? this.restore.delay;
                this.restore.speed = data.restore.speed ?? this.restore.speed;
                this.restore.follow = data.restore.follow ?? this.restore.follow;
            }
        }
    }

    const attractMode = "attract", minVelocityLengthSq = 0, minRestoreSpeed = 0.001, maxRestoreSpeed = 1, restoreEpsilon = 0.5;
    class Attractor extends pluginInteractivity.ExternalInteractorBase {
        handleClickMode;
        #interactedThisFrame;
        #maxDistance;
        #pluginManager;
        #restoreData;
        constructor(pluginManager, container) {
            super(container);
            this.#pluginManager = pluginManager;
            this.#maxDistance = 0;
            this.#interactedThisFrame = new Set();
            this.#restoreData = new Map();
            container.attract ??= { particles: [] };
            this.handleClickMode = (mode, interactivityData) => {
                const options = this.container.actualOptions, attract = options.interactivity?.modes.attract;
                if (!attract || mode !== attractMode) {
                    return;
                }
                container.attract ??= { particles: [] };
                container.attract.clicking = true;
                container.attract.count = 0;
                for (const particle of container.attract.particles) {
                    if (!this.isEnabled(interactivityData, particle)) {
                        continue;
                    }
                    particle.velocity.setTo(particle.initialVelocity);
                }
                container.attract.particles = [];
                container.attract.finish = false;
                setTimeout(() => {
                    if (container.destroyed) {
                        return;
                    }
                    container.attract ??= { particles: [] };
                    container.attract.clicking = false;
                }, attract.duration * engine.millisecondsToSeconds);
            };
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear() {
        }
        init() {
            const container = this.container, attract = container.actualOptions.interactivity?.modes.attract;
            if (!attract) {
                return;
            }
            this.#maxDistance = attract.distance;
            container.retina.attractModeDistance = attract.distance * container.retina.pixelRatio;
        }
        interact(interactivityData) {
            this.#interactedThisFrame.clear();
            const container = this.container, options = container.actualOptions, mouseMoveStatus = interactivityData.status === pluginInteractivity.mouseMoveEvent, events = options.interactivity?.events;
            if (!events) {
                return;
            }
            const { enable: hoverEnabled, mode: hoverMode } = events.onHover, { enable: clickEnabled, mode: clickMode } = events.onClick;
            if (mouseMoveStatus && hoverEnabled && engine.isInArray(attractMode, hoverMode)) {
                hoverAttract(this.#pluginManager, this.container, interactivityData, p => this.isEnabled(interactivityData, p), p => {
                    this.#trackInteractedParticle(p);
                });
            }
            else if (clickEnabled && engine.isInArray(attractMode, clickMode)) {
                clickAttract(this.#pluginManager, this.container, interactivityData, p => this.isEnabled(interactivityData, p), p => {
                    this.#trackInteractedParticle(p);
                });
            }
            this.#restoreParticles();
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
            if ((!mouse.position || !events?.onHover.enable) && (!mouse.clickPosition || !events?.onClick.enable)) {
                return false;
            }
            const hoverMode = events.onHover.mode, clickMode = events.onClick.mode;
            return engine.isInArray(attractMode, hoverMode) || engine.isInArray(attractMode, clickMode);
        }
        loadModeOptions(options, ...sources) {
            options.attract ??= new Attract();
            for (const source of sources) {
                options.attract.load(source?.attract);
            }
        }
        reset() {
        }
        #restoreParticles() {
            const restore = this.container.actualOptions.interactivity?.modes.attract?.restore;
            if (!restore?.enable || !this.#restoreData.size) {
                return;
            }
            const now = Date.now(), restoreDelay = restore.delay * engine.millisecondsToSeconds, restoreSpeed = Math.max(minRestoreSpeed, Math.min(maxRestoreSpeed, restore.speed));
            for (const [particle, restoreData] of this.#restoreData) {
                if (this.#interactedThisFrame.has(particle)) {
                    continue;
                }
                if (particle.destroyed) {
                    this.#restoreData.delete(particle);
                    continue;
                }
                const target = restoreData.target;
                if (now - restoreData.lastInteractionTime < restoreDelay) {
                    continue;
                }
                let dx = target.x - particle.position.x, dy = target.y - particle.position.y, dz = target.z - particle.position.z;
                if (restore.follow && particle.options.move.enable) {
                    const { x: vx, y: vy, z: vz } = particle.velocity, velocityLengthSq = vx * vx + vy * vy + vz * vz;
                    if (velocityLengthSq > minVelocityLengthSq) {
                        const parallelScale = (dx * vx + dy * vy + dz * vz) / velocityLengthSq;
                        dx -= vx * parallelScale;
                        dy -= vy * parallelScale;
                        dz -= vz * parallelScale;
                    }
                }
                particle.position.x += dx * restoreSpeed;
                particle.position.y += dy * restoreSpeed;
                particle.position.z += dz * restoreSpeed;
                if (Math.abs(dx) <= restoreEpsilon && Math.abs(dy) <= restoreEpsilon) {
                    particle.position.x = target.x;
                    particle.position.y = target.y;
                    particle.position.z = target.z;
                    this.#restoreData.delete(particle);
                    continue;
                }
            }
        }
        #trackInteractedParticle(particle) {
            this.#interactedThisFrame.add(particle);
            const restore = this.container.actualOptions.interactivity?.modes.attract?.restore;
            if (!restore?.enable) {
                return;
            }
            const now = Date.now();
            let restoreData = this.#restoreData.get(particle);
            if (!restoreData) {
                restoreData = {
                    target: particle.position.copy(),
                    lastInteractionTime: now,
                };
                this.#restoreData.set(particle, restoreData);
            }
            restoreData.lastInteractionTime = now;
        }
    }

    async function loadExternalAttractInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalAttract", container => {
                return Promise.resolve(new Attractor(e.pluginManager, container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalAttractInteraction = loadExternalAttractInteraction;

    exports.Attract = Attract;
    exports.loadExternalAttractInteraction = loadExternalAttractInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalAttractInteraction: (globalThis.__tsParticlesInternals.interactions.externalAttract || {}).loadExternalAttractInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
