// @tsparticles/interaction-external-destroy/esm/Destroyer.js
import { ExternalInteractorBase, isDivModeEnabled, mouseMoveEvent, } from "@tsparticles/plugin-interactivity";
import { isInArray } from "@tsparticles/engine";
import { divDestroy, mouseDestroy } from "./Utils.js";
import { Destroy } from "./Options/Classes/Destroy.js";
const destroyMode = "destroy";
export class Destroyer extends ExternalInteractorBase {
    #maxDistance;
    constructor(container) {
        super(container);
        this.#maxDistance = 0;
    }
    get maxDistance() {
        return this.#maxDistance;
    }
    clear() {
    }
    init() {
        const container = this.container, destroy = container.actualOptions.interactivity?.modes.destroy;
        if (!destroy) {
            return;
        }
        this.#maxDistance = destroy.distance;
        container.retina.destroyModeDistance = destroy.distance * container.retina.pixelRatio;
    }
    interact(interactivityData) {
        const container = this.container, options = container.actualOptions, events = options.interactivity?.events, mouseMoveStatus = interactivityData.status === mouseMoveEvent;
        if (!events) {
            return;
        }
        const hoverEnabled = events.onHover.enable, hoverMode = events.onHover.mode, divs = events.onDiv;
        if (mouseMoveStatus && hoverEnabled && isInArray(destroyMode, hoverMode)) {
            mouseDestroy(this.container, interactivityData);
        }
        else {
            divDestroy(this.container, divs, destroyMode);
        }
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, options = container.actualOptions, mouse = interactivityData.mouse, events = (particle?.interactivity ?? options.interactivity)?.events;
        if (!events) {
            return false;
        }
        const divs = events.onDiv;
        return ((!!mouse.position && events.onHover.enable && isInArray(destroyMode, events.onHover.mode)) ||
            isDivModeEnabled(destroyMode, divs));
    }
    loadModeOptions(options, ...sources) {
        options.destroy ??= new Destroy();
        for (const source of sources) {
            options.destroy.load(source?.destroy);
        }
    }
    reset() {
    }
}
