// nuxi/dist/dist-BkNIIgFa.mjs
//#region ../../node_modules/.pnpm/std-env@4.1.0/node_modules/std-env/dist/index.mjs
const e = globalThis.process?.env || Object.create(null), t = globalThis.process || { env: e }, n = t !== void 0 && t.env && t.env.NODE_ENV || void 0, r = [
	[`claude`, [`CLAUDECODE`, `CLAUDE_CODE`]],
	[`replit`, [`REPL_ID`]],
	[`gemini`, [`GEMINI_CLI`]],
	[`codex`, [`CODEX_SANDBOX`, `CODEX_THREAD_ID`]],
	[`opencode`, [`OPENCODE`]],
	[`pi`, [i(`PATH`, /\.pi[\\/]agent/)]],
	[`auggie`, [`AUGMENT_AGENT`]],
	[`goose`, [`GOOSE_PROVIDER`]],
	[`devin`, [i(`EDITOR`, /devin/)]],
	[`cursor`, [`CURSOR_AGENT`]],
	[`kiro`, [i(`TERM_PROGRAM`, /kiro/)]]
];
function i(t, n) {
	return () => {
		let r = e[t];
		return r ? n.test(r) : !1;
	};
}
function a() {
	let t = e.AI_AGENT;
	if (t) return { name: t.toLowerCase() };
	for (let [t, n] of r) for (let r of n) if (typeof r == `string` ? e[r] : r()) return { name: t };
	return {};
}
const o = a();
o.name;
const c = !!o.name, l = [
	[`APPVEYOR`],
	[
		`AWS_AMPLIFY`,
		`AWS_APP_ID`,
		{ ci: !0 }
	],
	[`AZURE_PIPELINES`, `SYSTEM_TEAMFOUNDATIONCOLLECTIONURI`],
	[`AZURE_STATIC`, `INPUT_AZURE_STATIC_WEB_APPS_API_TOKEN`],
	[`APPCIRCLE`, `AC_APPCIRCLE`],
	[`BAMBOO`, `bamboo_planKey`],
	[`BITBUCKET`, `BITBUCKET_COMMIT`],
	[`BITRISE`, `BITRISE_IO`],
	[`BUDDY`, `BUDDY_WORKSPACE_ID`],
	[`BUILDKITE`],
	[`CIRCLE`, `CIRCLECI`],
	[`CIRRUS`, `CIRRUS_CI`],
	[
		`CLOUDFLARE_PAGES`,
		`CF_PAGES`,
		{ ci: !0 }
	],
	[
		`CLOUDFLARE_WORKERS`,
		`WORKERS_CI`,
		{ ci: !0 }
	],
	[`GOOGLE_CLOUDRUN`, `K_SERVICE`],
	[`GOOGLE_CLOUDRUN_JOB`, `CLOUD_RUN_JOB`],
	[`CODEBUILD`, `CODEBUILD_BUILD_ARN`],
	[`CODEFRESH`, `CF_BUILD_ID`],
	[`DRONE`],
	[`DRONE`, `DRONE_BUILD_EVENT`],
	[`DSARI`],
	[`GITHUB_ACTIONS`],
	[`GITLAB`, `GITLAB_CI`],
	[`GITLAB`, `CI_MERGE_REQUEST_ID`],
	[`GOCD`, `GO_PIPELINE_LABEL`],
	[`LAYERCI`],
	[`JENKINS`, `JENKINS_URL`],
	[`HUDSON`, `HUDSON_URL`],
	[`MAGNUM`],
	[`NETLIFY`],
	[
		`NETLIFY`,
		`NETLIFY_LOCAL`,
		{ ci: !1 }
	],
	[`NEVERCODE`],
	[`RENDER`],
	[`SAIL`, `SAILCI`],
	[`SEMAPHORE`],
	[`SCREWDRIVER`],
	[`SHIPPABLE`],
	[`SOLANO`, `TDDIUM`],
	[`STRIDER`],
	[`TEAMCITY`, `TEAMCITY_VERSION`],
	[`TRAVIS`],
	[`VERCEL`, `NOW_BUILDER`],
	[
		`VERCEL`,
		`VERCEL`,
		{ ci: !1 }
	],
	[
		`VERCEL`,
		`VERCEL_ENV`,
		{ ci: !1 }
	],
	[`APPCENTER`, `APPCENTER_BUILD_ID`],
	[
		`CODESANDBOX`,
		`CODESANDBOX_SSE`,
		{ ci: !1 }
	],
	[
		`CODESANDBOX`,
		`CODESANDBOX_HOST`,
		{ ci: !1 }
	],
	[`STACKBLITZ`],
	[`STORMKIT`],
	[`CLEAVR`],
	[`ZEABUR`],
	[
		`CODESPHERE`,
		`CODESPHERE_APP_ID`,
		{ ci: !0 }
	],
	[`RAILWAY`, `RAILWAY_PROJECT_ID`],
	[`RAILWAY`, `RAILWAY_SERVICE_ID`],
	[`DENO-DEPLOY`, `DENO_DEPLOY`],
	[`DENO-DEPLOY`, `DENO_DEPLOYMENT_ID`],
	[
		`FIREBASE_APP_HOSTING`,
		`FIREBASE_APP_HOSTING`,
		{ ci: !0 }
	],
	[
		`EDGEONE_PAGES`,
		`EO_PAGES_CI`,
		{ ci: !0 }
	]
];
function u() {
	for (let t of l) if (e[t[1] || t[0]]) return {
		name: t[0].toLowerCase(),
		...t[2]
	};
	return e.SHELL === `/bin/jsh` && t.versions?.webcontainer ? {
		name: `stackblitz`,
		ci: !1
	} : {
		name: ``,
		ci: !1
	};
}
const d = u(), f = d.name, p = t.platform || ``, m = !!e.CI || d.ci !== !1, h = !!t.stdout?.isTTY;
e.DEBUG;
const v = n === `test` || !!e.TEST;
n === `production` || e.MODE;
n === `dev` || n === `development` || e.MODE;
const x = !!e.MINIMAL || m || v || !h, S = /^win/i.test(p);
/^linux/i.test(p);
/^darwin/i.test(p);
!e.NO_COLOR && (e.FORCE_COLOR || (h || S) && e.TERM);
const E = (t.versions?.node || ``).replace(/^v/, ``) || null;
Number(E?.split(`.`)[0]);
const O = !!t?.versions?.node, k = `Bun` in globalThis, A = `Deno` in globalThis, j = `fastly` in globalThis, M = `Netlify` in globalThis, N = `EdgeRuntime` in globalThis, P = globalThis.navigator?.userAgent === `Cloudflare-Workers`, F = [
	[M, `netlify`],
	[N, `edge-light`],
	[P, `workerd`],
	[j, `fastly`],
	[A, `deno`],
	[k, `bun`],
	[O, `node`]
];
function I() {
	let e = F.find((e) => e[0]);
	if (e) return { name: e[1] };
}
I()?.name;
//#endregion
export { k as a, h as i, c as n, v as o, f as r, x as s, A as t };
