// @tsparticles/interaction-external-slow/tsparticles.interaction.external.slow.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalSlow = global.__tsParticlesInternals.interactions.externalSlow || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    class Slow {
        factor;
        radius;
        constructor() {
            this.factor = 3;
            this.radius = 200;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.factor !== undefined) {
                this.factor = data.factor;
            }
            if (data.radius !== undefined) {
                this.radius = data.radius;
            }
        }
    }

    const slowMode = "slow", minRadius = 0;
    class Slower extends pluginInteractivity.ExternalInteractorBase {
        #maxDistance;
        constructor(container) {
            super(container);
            this.#maxDistance = 0;
        }
        get maxDistance() {
            return this.#maxDistance;
        }
        clear(particle, _delta, force) {
            if (particle.slow.inRange && !force) {
                return;
            }
            particle.slow.factor = 1;
        }
        init() {
            const container = this.container, slow = container.actualOptions.interactivity?.modes.slow;
            if (!slow) {
                return;
            }
            this.#maxDistance = slow.radius;
            container.retina.slowModeRadius = slow.radius * container.retina.pixelRatio;
        }
        interact() {
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
            return !!events?.onHover.enable && !!mouse.position && engine.isInArray(slowMode, events.onHover.mode);
        }
        loadModeOptions(options, ...sources) {
            options.slow ??= new Slow();
            for (const source of sources) {
                options.slow.load(source?.slow);
            }
        }
        reset(interactivityData, particle) {
            particle.slow.inRange = false;
            const container = this.container, options = container.actualOptions, mousePos = interactivityData.mouse.position, radius = container.retina.slowModeRadius, slowOptions = options.interactivity?.modes.slow;
            if (!slowOptions || !radius || radius < minRadius || !mousePos) {
                return;
            }
            const particlePos = particle.getPosition(), dist = engine.getDistance(mousePos, particlePos), proximityFactor = dist / radius, slowFactor = slowOptions.factor, { slow } = particle;
            if (dist > radius) {
                return;
            }
            slow.inRange = true;
            slow.factor = proximityFactor / slowFactor;
        }
    }

    async function loadExternalSlowInteraction(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalSlow", container => {
                return Promise.resolve(new Slower(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalSlowInteraction = loadExternalSlowInteraction;

    exports.Slow = Slow;
    exports.loadExternalSlowInteraction = loadExternalSlowInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalSlowInteraction: (globalThis.__tsParticlesInternals.interactions.externalSlow || {}).loadExternalSlowInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
