// @tsparticles/updater-opacity/browser/OpacityUpdater.js
import { getRandom, getRangeValue, initParticleNumericAnimationValue, percentDenominator, updateAnimation, } from "@tsparticles/engine";
import { Opacity } from "./Options/Classes/Opacity.js";
export class OpacityUpdater {
    #container;
    constructor(container) {
        this.#container = container;
    }
    init(particle) {
        const opacityOptions = particle.options.opacity, pxRatio = 1;
        if (!opacityOptions) {
            return;
        }
        particle.opacity = initParticleNumericAnimationValue(opacityOptions, pxRatio);
        const opacityAnimation = opacityOptions.animation;
        if (opacityAnimation.enable) {
            particle.opacity.velocity =
                (getRangeValue(opacityAnimation.speed) / percentDenominator) * this.#container.retina.reduceFactor;
            if (!opacityAnimation.sync) {
                particle.opacity.velocity *= getRandom();
            }
        }
    }
    isEnabled(particle) {
        const none = 0;
        return (!particle.destroyed &&
            !particle.spawning &&
            !!particle.opacity &&
            particle.opacity.enable &&
            ((particle.opacity.maxLoops ?? none) <= none ||
                ((particle.opacity.maxLoops ?? none) > none &&
                    (particle.opacity.loops ?? none) < (particle.opacity.maxLoops ?? none))));
    }
    loadOptions(options, ...sources) {
        options.opacity ??= new Opacity();
        for (const source of sources) {
            options.opacity.load(source?.opacity);
        }
    }
    reset(particle) {
        if (!particle.opacity) {
            return;
        }
        particle.opacity.time = 0;
        particle.opacity.loops = 0;
    }
    update(particle, delta) {
        if (!this.isEnabled(particle) || !particle.opacity || !particle.options.opacity) {
            return;
        }
        updateAnimation(particle, particle.opacity, true, particle.options.opacity.animation.destroy, delta);
    }
}
