// @tsparticles/plugin-rgb-color/tsparticles.plugin.rgbColor.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.1.3 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.rgbColor = global.__tsParticlesInternals.plugins.rgbColor || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    var RgbIndexes;
    (function (RgbIndexes) {
        RgbIndexes[RgbIndexes["r"] = 1] = "r";
        RgbIndexes[RgbIndexes["g"] = 2] = "g";
        RgbIndexes[RgbIndexes["b"] = 3] = "b";
        RgbIndexes[RgbIndexes["a"] = 5] = "a";
    })(RgbIndexes || (RgbIndexes = {}));
    const rgbRegex = /rgba?\(\s*(\d{1,3})\s*[\s,]\s*(\d{1,3})\s*[\s,]\s*(\d{1,3})\s*([\s,]\s*(0|1|0?\.\d+|(\d{1,3})%)\s*)?\)/i;
    class RgbColorManager {
        accepts(input) {
            return input.startsWith("rgb");
        }
        handleColor(color) {
            const colorValue = color.value, rgbColor = colorValue.rgb ?? color.value;
            if (!("r" in rgbColor) || !("g" in rgbColor) || !("b" in rgbColor)) {
                return;
            }
            return rgbColor;
        }
        handleRangeColor(color) {
            const colorValue = color.value, rgbColor = colorValue.rgb ?? color.value;
            if (!("r" in rgbColor) || !("g" in rgbColor) || !("b" in rgbColor)) {
                return;
            }
            return {
                r: engine.getRangeValue(rgbColor.r),
                g: engine.getRangeValue(rgbColor.g),
                b: engine.getRangeValue(rgbColor.b),
            };
        }
        parseString(input) {
            if (!this.accepts(input)) {
                return;
            }
            const result = rgbRegex.exec(input), radix = 10, minLength = 4, defaultAlpha = 1;
            return result
                ? {
                    a: result.length > minLength ? engine.parseAlpha(result[RgbIndexes.a]) : defaultAlpha,
                    b: parseInt(result[RgbIndexes.b] ?? "0", radix),
                    g: parseInt(result[RgbIndexes.g] ?? "0", radix),
                    r: parseInt(result[RgbIndexes.r] ?? "0", radix),
                }
                : undefined;
        }
    }

    async function loadRgbColorPlugin(engine) {
        engine.checkVersion("4.1.3");
        await engine.pluginManager.register(e => {
            e.pluginManager.addColorManager("rgb", new RgbColorManager());
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadRgbColorPlugin = loadRgbColorPlugin;

    exports.loadRgbColorPlugin = loadRgbColorPlugin;

}));
Object.assign(globalThis.window || globalThis, { loadRgbColorPlugin: (globalThis.__tsParticlesInternals.plugins.rgbColor || {}).loadRgbColorPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
