// @tsparticles/shape-image/cjs/index.lazy.js
import { shapeTypes } from "./Utils.js";
const extLength = 3;
function addLoadImageToEngine(engine) {
    engine.getImages ??= (container) => {
        engine.images ??= new Map();
        let images = engine.images.get(container);
        if (!images) {
            images = [];
            engine.images.set(container, images);
        }
        return images;
    };
    engine.loadImage ??= async (container, data) => {
        if (!engine.getImages) {
            throw new Error("No images collection found");
        }
        if (!data.name && !data.src) {
            throw new Error("No image source provided");
        }
        engine.images ??= new Map();
        const containerImages = engine.getImages(container);
        if (containerImages.some((t) => t.name === data.name || t.source === data.src)) {
            return;
        }
        try {
            const image = {
                gif: data.gif,
                name: data.name ?? data.src,
                source: data.src,
                type: data.src.substring(data.src.length - extLength),
                error: false,
                loading: true,
                replaceColor: data.replaceColor,
                ratio: data.width && data.height ? data.width / data.height : undefined,
            };
            containerImages.push(image);
            engine.images.set(container, containerImages);
            let imageFunc;
            if (data.gif) {
                const { loadGifImage } = await import("./GifUtils/Utils.js");
                imageFunc = (img) => loadGifImage(img, { colorSpace: "srgb" });
            }
            else if (data.replaceColor) {
                const { downloadSvgImage } = await import("./Utils.js");
                imageFunc = downloadSvgImage;
            }
            else {
                const { loadImage } = await import("./Utils.js");
                imageFunc = loadImage;
            }
            await imageFunc(image);
        }
        catch {
            throw new Error(`${data.name ?? data.src} not found`);
        }
    };
}
export async function loadImageShape(engine) {
    engine.checkVersion("4.1.3");
    await engine.pluginManager.register(async (e) => {
        const { ImagePreloaderPlugin } = await import("./ImagePreloader.js");
        addLoadImageToEngine(e);
        e.pluginManager.addPlugin(new ImagePreloaderPlugin(e));
        e.pluginManager.addShape(shapeTypes, async (container) => {
            const { ImageDrawer } = await import("./ImageDrawer.js");
            return new ImageDrawer(e, container);
        });
    });
}
