// @tsparticles/engine/browser/Enums/Types/EasingType.js
export var EasingType;
(function (EasingType) {
    EasingType["easeInBack"] = "ease-in-back";
    EasingType["easeInBounce"] = "ease-in-bounce";
    EasingType["easeInCirc"] = "ease-in-circ";
    EasingType["easeInCubic"] = "ease-in-cubic";
    EasingType["easeInElastic"] = "ease-in-elastic";
    EasingType["easeInExpo"] = "ease-in-expo";
    EasingType["easeInGaussian"] = "ease-in-gaussian";
    EasingType["easeInLinear"] = "ease-in-linear";
    EasingType["easeInQuad"] = "ease-in-quad";
    EasingType["easeInQuart"] = "ease-in-quart";
    EasingType["easeInQuint"] = "ease-in-quint";
    EasingType["easeInSigmoid"] = "ease-in-sigmoid";
    EasingType["easeInSine"] = "ease-in-sine";
    EasingType["easeInSmoothstep"] = "ease-in-smoothstep";
    EasingType["easeOutBack"] = "ease-out-back";
    EasingType["easeOutBounce"] = "ease-out-bounce";
    EasingType["easeOutCirc"] = "ease-out-circ";
    EasingType["easeOutCubic"] = "ease-out-cubic";
    EasingType["easeOutElastic"] = "ease-out-elastic";
    EasingType["easeOutExpo"] = "ease-out-expo";
    EasingType["easeOutGaussian"] = "ease-out-gaussian";
    EasingType["easeOutLinear"] = "ease-out-linear";
    EasingType["easeOutQuad"] = "ease-out-quad";
    EasingType["easeOutQuart"] = "ease-out-quart";
    EasingType["easeOutQuint"] = "ease-out-quint";
    EasingType["easeOutSigmoid"] = "ease-out-sigmoid";
    EasingType["easeOutSine"] = "ease-out-sine";
    EasingType["easeOutSmoothstep"] = "ease-out-smoothstep";
    EasingType["easeInOutBack"] = "ease-in-out-back";
    EasingType["easeInOutBounce"] = "ease-in-out-bounce";
    EasingType["easeInOutCirc"] = "ease-in-out-circ";
    EasingType["easeInOutCubic"] = "ease-in-out-cubic";
    EasingType["easeInOutElastic"] = "ease-in-out-elastic";
    EasingType["easeInOutExpo"] = "ease-in-out-expo";
    EasingType["easeInOutGaussian"] = "ease-in-out-gaussian";
    EasingType["easeInOutLinear"] = "ease-in-out-linear";
    EasingType["easeInOutQuad"] = "ease-in-out-quad";
    EasingType["easeInOutQuart"] = "ease-in-out-quart";
    EasingType["easeInOutQuint"] = "ease-in-out-quint";
    EasingType["easeInOutSigmoid"] = "ease-in-out-sigmoid";
    EasingType["easeInOutSine"] = "ease-in-out-sine";
    EasingType["easeInOutSmoothstep"] = "ease-in-out-smoothstep";
})(EasingType || (EasingType = {}));
